{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "757b3d6c",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9a417da1",
   "metadata": {},
   "outputs": [],
   "source": [
    "#final inshaallh\n",
    "from pymongo import MongoClient\n",
    "import tkinter as tk\n",
    "from tkinter import messagebox\n",
    "from pymongo import DESCENDING\n",
    "import hashlib\n",
    "import csv\n",
    "from pymongo import MongoClient\n",
    "from tkinter import ttk\n",
    "import subprocess\n",
    "import tempfile\n",
    "\n",
    "import tkinter as tk\n",
    "from tkinter import ttk\n",
    "class App:\n",
    "    def __init__(self):\n",
    "        self.hello = None        \n",
    "        self.window = None\n",
    "        self.hello_update= None\n",
    "        # Connect to MongoDB\n",
    "        self.client = MongoClient('mongodb://localhost:27017')\n",
    "        self.db = self.client['breast']\n",
    "        self.collection =self.db['cancerf'] \n",
    "        self.collection2 =self.db['clinical_data'] \n",
    "\n",
    "        \n",
    "        \n",
    "###########\n",
    "###########\n",
    "########### page 1 :\n",
    "\n",
    "    def open_chcollection(self):\n",
    "        chcollection = tk.Toplevel()\n",
    "        chcollection.title(\"Collection\")\n",
    "        chcollection.geometry(\"600x400\")\n",
    "        chcollection.configure(bg=\"gray\")\n",
    "        chcollection.attributes(\"-fullscreen\", True)\n",
    "        screen_width = chcollection.winfo_screenwidth()\n",
    "        screen_height = chcollection.winfo_screenheight()\n",
    "        \n",
    "        canvas = tk.Canvas(chcollection, width=screen_width, height=screen_height)\n",
    "        canvas.place(x=0, y=0)\n",
    "\n",
    "        # Define the pastel gradient colors\n",
    "        gradient_colors = [\"#F8B195\", \"#F67280\", \"#C06C84\", \"#6C5B7B\", \"#355C7D\"]\n",
    "        for i in range(len(gradient_colors)):\n",
    "            color = gradient_colors[i]\n",
    "            x0 = int((i / len(gradient_colors)) * screen_width)\n",
    "            x1 = int(((i + 1) / len(gradient_colors)) * screen_width)\n",
    "            canvas.create_rectangle(x0, 0, x1, screen_height, fill=color, width=0)\n",
    "            \n",
    "        # choose collection back button\n",
    "        button3 = tk.Button(chcollection, text=\"Back\", font=(\"Copperplate\", 25, \"bold\"), fg=\"#F6B195\", bg=\"black\", command=chcollection.destroy)\n",
    "        button3.place(x=1700, y=1000, width=200, height=70)\n",
    "     \n",
    "     ####################################################################################################\n",
    "        \n",
    "        def open_query_cancer():\n",
    "            \n",
    "## => => => => => => => => => => => => => => => @labels & textboxes & buttons@ => => => => => => => => => => => => => => => \n",
    "\n",
    "            query_cancer = tk.Toplevel()\n",
    "            query_cancer.title(\"Make queries on <Cancer collection>\")\n",
    "            query_cancer.geometry(\"600x400\")\n",
    "            query_cancer.configure(bg=\"gray\")\n",
    "            query_cancer.attributes(\"-fullscreen\", True)\n",
    "            screen_width = query_cancer.winfo_screenwidth()\n",
    "            screen_height = query_cancer.winfo_screenheight()\n",
    "\n",
    "            canvas = tk.Canvas(query_cancer, width=screen_width, height=screen_height)\n",
    "            canvas.place(x=0, y=0)\n",
    "            \n",
    "\n",
    "            gradient_colors = [\"#F8B195\", \"#F67280\", \"#C06C84\", \"#6C5B7B\", \"#355C7D\"]\n",
    "            for i in range(len(gradient_colors)):\n",
    "                color = gradient_colors[i]\n",
    "                x0 = int((i / len(gradient_colors)) * screen_width)\n",
    "                x1 = int(((i + 1) / len(gradient_colors)) * screen_width)\n",
    "                canvas.create_rectangle(x0, 0, x1, screen_height, fill=color, width=0)\n",
    "            buttonbb = tk.Button(query_cancer, text=\"Back\", font=(\"Copperplate\", 10, \"bold\"), fg=\"#F6B195\", bg=\"black\", command=query_cancer.destroy)\n",
    "            buttonbb.place(x=0,y=0, width=100, height=40)\n",
    "            \n",
    "            label2 = tk.Label(query_cancer, text=\"Label 2\", anchor=\"center\", cursor=\"star\")\n",
    "            label2.place(x=100, y=0, width=1900, height=40)\n",
    "            label2.config(text=\"Big data cancer file queries\", anchor=\"center\", cursor=\"star\", bg=\"black\",\n",
    "                              font=(\"Copperplate\", 15, \"bold\"), fg=\"#F6B195\")\n",
    "\n",
    "            # Entry boxes for inserting data\n",
    "            \n",
    "# Entry boxes for inserting data\n",
    "\n",
    "\n",
    "            ###########################################################################3\n",
    "            label_id = tk.Label(query_cancer, text=\"ID:\", width=10, height=2, font=(\"Copperplate\", 10, \"bold\"), fg=\"white\", bg=\"black\", borderwidth=2, relief=\"sunken\")\n",
    "            label_id.place(x=50, y=50)\n",
    "\n",
    "            self.entry_id = tk.Entry(query_cancer)\n",
    "            self.entry_id.place(x=150, y=60)\n",
    "\n",
    "            label_symbol = tk.Label(query_cancer, text=\"Gene Symbol:\",width=12, height=2, font=(\"Copperplate\", 10, \"bold\"), fg=\"white\", bg=\"black\")\n",
    "            label_symbol.place(x=50, y=90)\n",
    "            self.entry_symbol = tk.Entry(query_cancer)\n",
    "            self.entry_symbol.place(x=150, y=100)\n",
    "\n",
    "            label_name = tk.Label(query_cancer, text=\"Gene Name:\",width=10, height=2, font=(\"Copperplate\", 10, \"bold\"), fg=\"white\", bg=\"black\")\n",
    "            label_name.place(x=50, y=130)\n",
    "            self.entry_name = tk.Entry(query_cancer)\n",
    "            self.entry_name.place(x=150, y=140)\n",
    "\n",
    "            label_s1 = tk.Label(query_cancer, text=\"Sample 1:\",width=10, height=2, font=(\"Copperplate\", 10, \"bold\"), fg=\"white\", bg=\"black\")\n",
    "            label_s1.place(x=50, y=170)\n",
    "            self.entry_s1 = tk.Entry(query_cancer)\n",
    "            self.entry_s1.place(x=150, y=180)\n",
    "\n",
    "            label_s2 = tk.Label(query_cancer, text=\"sample 2:\",width=10, height=2, font=(\"Copperplate\", 10, \"bold\"), fg=\"white\", bg=\"black\")\n",
    "            label_s2.place(x=50, y=210)\n",
    "            self.entry_s2 = tk.Entry(query_cancer)\n",
    "            self.entry_s2.place(x=150, y=220)\n",
    "\n",
    "            label_s3 = tk.Label(query_cancer, text=\"Sample 3:\",width=10, height=2, font=(\"Copperplate\", 10, \"bold\"), fg=\"white\", bg=\"black\")\n",
    "            label_s3.place(x=50, y=250)\n",
    "            self.entry_s3 = tk.Entry(query_cancer)\n",
    "            self.entry_s3.place(x=150, y=260)\n",
    "\n",
    "            label_s4 = tk.Label(query_cancer, text=\"Sample 4:\",width=10, height=2, font=(\"Copperplate\", 10, \"bold\"), fg=\"white\", bg=\"black\")\n",
    "            label_s4.place(x=50, y=325)\n",
    "            self.entry_s4 = tk.Entry(query_cancer)\n",
    "            self.entry_s4.place(x=150, y=330)\n",
    "\n",
    "            label_s5 = tk.Label(query_cancer, text=\"Sample 5: \",width=10, height=2, font=(\"Copperplate\", 10, \"bold\"), fg=\"white\", bg=\"black\")\n",
    "            label_s5.place(x=50, y=365)\n",
    "            self.entry_s5 = tk.Entry(query_cancer)\n",
    "            self.entry_s5.place(x=150, y=370)\n",
    "\n",
    "            label_s6 = tk.Label(query_cancer, text=\"Sample 6:\",width=10, height=2, font=(\"Copperplate\", 10, \"bold\"), fg=\"white\", bg=\"black\")\n",
    "            label_s6.place(x=50, y=405)\n",
    "            self.entry_s6 = tk.Entry(query_cancer)\n",
    "            self.entry_s6.place(x=150, y=410)\n",
    "\n",
    "            label_s7 = tk.Label(query_cancer, text=\"Sample 7:\",width=10, height=2, font=(\"Copperplate\", 10, \"bold\"), fg=\"white\", bg=\"black\")\n",
    "            label_s7.place(x=50, y=445)\n",
    "            self.entry_s7 = tk.Entry(query_cancer)\n",
    "            self.entry_s7.place(x=150, y=450)\n",
    "\n",
    "            label_s8 = tk.Label(query_cancer, text=\"Sample 8:\",width=10, height=2, font=(\"Copperplate\", 10, \"bold\"), fg=\"white\", bg=\"black\")\n",
    "            label_s8.place(x=50, y=485)\n",
    "            self.entry_s8 = tk.Entry(query_cancer)\n",
    "            self.entry_s8.place(x=150, y=490)\n",
    "\n",
    "            label_s9 = tk.Label(query_cancer, text=\"sample 9:\",width=10, height=2, font=(\"Copperplate\", 10, \"bold\"), fg=\"white\", bg=\"black\")\n",
    "            label_s9.place(x=50, y=525)\n",
    "            self.entry_s9 = tk.Entry(query_cancer)\n",
    "            self.entry_s9.place(x=150, y=530)\n",
    "\n",
    "            label_s10 = tk.Label(query_cancer, text=\"Sample 10:\",width=10, height=2, font=(\"Copperplate\", 10, \"bold\"), fg=\"white\", bg=\"black\")\n",
    "            label_s10.place(x=50, y=565)\n",
    "            self.entry_s10 = tk.Entry(query_cancer)\n",
    "            self.entry_s10.place(x=150, y=570)\n",
    "            \n",
    "\n",
    "            insert_button = tk.Button(query_cancer, text=\"Insert\", command=self.insert_data, width=10, height=2, font=(\"Copperplate\", 25, \"bold\"), fg=\"#F6B195\", bg=\"black\")\n",
    "            insert_button.place(x=50, y=650)\n",
    "\n",
    "            update_button = tk.Button(query_cancer, text=\"Update\",command=self.update_document, width=10, height=2, font=(\"Copperplate\", 25, \"bold\"), fg=\"#F6B195\", bg=\"black\")\n",
    "            update_button.place(x=800, y=600)\n",
    "            \n",
    " \n",
    "            delete_button = tk.Button(query_cancer, text=\"Delete\", command=self.delete_data,width=10, height=2, font=(\"Copperplate\", 25, \"bold\"), fg=\"#F6B195\", bg=\"black\")\n",
    "            delete_button.place(x=800, y=200)\n",
    "            \n",
    "            calculate_AVG_button = tk.Button(query_cancer, text=\" Calculate Average for \\n all sample recoreds\", command=self.calculate_avg ,width=20, height=2, font=(\"Copperplate\", 15, \"bold\"), fg=\"#F6B195\", bg=\"black\")\n",
    "            calculate_AVG_button.place(x=400, y=450)\n",
    "            \n",
    "            \n",
    "################ $in operator\n",
    "            #get gene symbol for all <less>  than 1\n",
    "            sep_in = tk.Label(query_cancer, text=\"  Return the 3 accession doument\\n usin $in operator :\",width=52, height=2, font=(\"Copperplate\", 10, \"bold\"), fg=\"#F6B195\", bg=\"black\",anchor='w')\n",
    "            sep_in.place(x=1535, y=40)\n",
    "\n",
    "            self.accession1_label = tk.Label(query_cancer, text=\"Accession 1:\",width=15, height=2, font=(\"Copperplate\", 10, \"bold\"), fg=\"white\", bg=\"black\")\n",
    "            self.accession1_label.place(x=1540, y=100)           # sep_gt.place(x=1535, y=40)\n",
    "\n",
    "            self.accession1_entry = tk.Entry(query_cancer)\n",
    "            self.accession1_entry.place(x=1540, y=130)\n",
    "\n",
    "            self.accession2_label = tk.Label(query_cancer, text=\"Accession 2:\",width=15, height=2, font=(\"Copperplate\", 10, \"bold\"), fg=\"white\", bg=\"black\")\n",
    "            self.accession2_label.place(x=1540, y=170)\n",
    "\n",
    "            self.accession2_entry = tk.Entry(query_cancer)\n",
    "            self.accession2_entry.place(x=1540, y=200)\n",
    "\n",
    "            self.accession3_label = tk.Label(query_cancer, text=\"Accession 3:\",width=15, height=2, font=(\"Copperplate\", 10, \"bold\"), fg=\"white\", bg=\"black\")\n",
    "            self.accession3_label.place(x=1540, y=240)\n",
    "\n",
    "            self.accession3_entry = tk.Entry(query_cancer)\n",
    "            self.accession3_entry.place(x=1540, y=270)\n",
    "\n",
    "            # Create button to trigger retrieval\n",
    "            self.retrieve_button = tk.Button(query_cancer, text=\" Get three records \", command=self.in_operator, width=20, height=2, font=(\"Copperplate\", 15, \"bold\"), fg=\"#F6B195\", bg=\"black\",anchor='w')\n",
    "            self.retrieve_button.place(x=1500, y=320)\n",
    "############## delete many \n",
    "\n",
    "\n",
    "#get gene symbol for all greater than 1\n",
    "            \n",
    "            #labels\n",
    "            sep_gt = tk.Label(query_cancer, text=\" Get the symbol geater than 1 :\",width=52, height=2, font=(\"Copperplate\", 10, \"bold\"), fg=\"#F6B195\", bg=\"black\",anchor='w')\n",
    "            sep_gt.place(x=400, y=40)\n",
    "        \n",
    "            db_name_symbol = tk.Label(query_cancer, text=\"gt_Database Name:\",width=25, height=2, font=(\"Copperplate\", 10, \"bold\"), fg=\"white\", bg=\"black\")\n",
    "            db_name_symbol.place(x=400, y=70)\n",
    "\n",
    "            collection_name_symbol = tk.Label(query_cancer, text=\"gt_Collection Name:\",width=25, height=2, font=(\"Copperplate\", 10, \"bold\"), fg=\"white\", bg=\"black\")\n",
    "            collection_name_symbol.place(x=400, y=90)\n",
    "\n",
    "            object_name_symbol = tk.Label(query_cancer, text=\"gt_Object Name:\",width=25, height=2, font=(\"Copperplate\", 10, \"bold\"), fg=\"white\", bg=\"black\")\n",
    "            object_name_symbol.place(x=400, y=110)\n",
    "\n",
    "            #entries 1\n",
    "            self.db_name = tk.Entry(query_cancer)\n",
    "            self.db_name.place(x=600, y=70)\n",
    "\n",
    "            self.collection_name = tk.Entry(query_cancer)\n",
    "            self.collection_name.place(x=600, y=90)\n",
    "\n",
    "            self.object_name=tk.Entry(query_cancer)\n",
    "            self.object_name.place(x=600, y=110)\n",
    "            \n",
    "            get_gene_symbols_button = tk.Button(query_cancer, text=\"Get Gene Symbols <gt>\", command=self.get_gene_symbols_with_greater_than_one,width=20, height=2, font=(\"Copperplate\", 20, \"bold\"), fg=\"#F6B195\", bg=\"black\")\n",
    "            get_gene_symbols_button.place(x=400, y=150)\n",
    "##########\n",
    "# DELETE MANY\n",
    "            self.field_label = tk.Label(query_cancer, text=\"Field:\",width=15, height=2, font=(\"Copperplate\", 10, \"bold\"), fg=\"white\", bg=\"black\")\n",
    "            self.field_label.place(x=800, y=830)\n",
    "\n",
    "            # Create a textbox for field input\n",
    "            self.field_entry = tk.Entry(query_cancer)\n",
    "            self.field_entry.place(x=800, y=860)\n",
    "\n",
    "            # Create a label for the value input\n",
    "            self.value_label = tk.Label(query_cancer, text=\"Value:\",width=15, height=2, font=(\"Copperplate\", 10, \"bold\"), fg=\"white\", bg=\"black\")\n",
    "            self.value_label.place(x=800, y=890)\n",
    "\n",
    "            # Create a textbox for value input\n",
    "            self.value_entry = tk.Entry(query_cancer)\n",
    "            self.value_entry.place(x=800, y=920)\n",
    "            delete_button = tk.Button(query_cancer, text=\"Delete Many Documents\", command=self.delete_Many_documents, width=20, height=2, font=(\"Copperplate\", 15, \"bold\"), fg=\"#F6B195\", bg=\"black\",anchor='w')\n",
    "            delete_button.place(x=800, y=950)\n",
    "            ###########################################################################\n",
    "            \n",
    "            #get gene symbol for all <less>  than 1\n",
    "            sep_gt = tk.Label(query_cancer, text=\" Get the symbol less than 1 :\",width=55, height=2, font=(\"Copperplate\", 10, \"bold\"), fg=\"#F6B195\", bg=\"black\",anchor='w')\n",
    "            sep_gt.place(x=1150, y=40)\n",
    "\n",
    "            db_name_label_less = tk.Label(query_cancer, text=\"Database Name:\",width=25, height=2, font=(\"Copperplate\", 10, \"bold\"), fg=\"white\", bg=\"black\")\n",
    "            db_name_label_less.place(x=1150, y=100)\n",
    "            \n",
    "            self.db_name_entry_less = tk.Entry(query_cancer)\n",
    "            self.db_name_entry_less.place(x=1350, y=100)\n",
    "\n",
    "          \n",
    "            object_name_label_less = tk.Label(query_cancer, text=\"Object Name:\",width=25, height=2, font=(\"Copperplate\", 10, \"bold\"), fg=\"white\", bg=\"black\")\n",
    "            object_name_label_less.place(x=1150, y=120)\n",
    "            \n",
    "            self.object_name_entry_less = tk.Entry(query_cancer)\n",
    "            self.object_name_entry_less.place(x=1350, y=120)\n",
    "           \n",
    "            collection_name_label_less = tk.Label(query_cancer, text=\"Collection Name:\",width=25, height=2, font=(\"Copperplate\", 10, \"bold\"), fg=\"white\", bg=\"black\")\n",
    "            collection_name_label_less.place(x=1150, y=140)\n",
    "\n",
    "            self.collection_name_entry_less = tk.Entry(query_cancer)\n",
    "            self.collection_name_entry_less.place(x=1350, y=140)\n",
    "            #entries\n",
    "\n",
    "                      \n",
    "            get_gene_name_button_less = tk.Button(query_cancer, text=\"Get Gene Symbols\", command=self.get_gene_symbols_with_less_than_one , width=10, height=2, font=(\"Copperplate\", 15, \"bold\"), fg=\"#F6B195\", bg=\"black\")\n",
    "            get_gene_name_button_less.place(x=1160, y=200,width=350, height=90)\n",
    "           \n",
    "            \n",
    "\n",
    "            ###########################################################################\n",
    "            # hash labels ,entries\n",
    "            sep = tk.Label(query_cancer, text=\"Hashing accession \\n number\",width=54, height=2,font=(\"Copperplate\", 10, \"bold\"), fg=\"#F6B195\", bg=\"black\")\n",
    "            sep.place(x=385, y=350)\n",
    "            hash_label = tk.Label(query_cancer, text=\"Enter ID :\",width=25, height=2, font=(\"Copperplate\", 10, \"bold\"), fg=\"white\", bg=\"black\")\n",
    "            hash_label.place(x=385, y=400)\n",
    "            \n",
    "            self.hash_entry = tk.Entry(query_cancer)\n",
    "            self.hash_entry.place(x=600, y=410)\n",
    "            \n",
    "            # Create a button to perform the hashing and database update\n",
    "            button = tk.Button(query_cancer, text=\"Hash\", command=self.hash_document_id,width=20, height=2, font=(\"Copperplate\", 20, \"bold\"), fg=\"#F6B195\", bg=\"black\")\n",
    "            button.place(x=385, y=450)\n",
    "\n",
    "           #######################################################################\n",
    "            #REFRENCE TO CLINICAL DATA STUFFS\n",
    "            refrence_label = tk.Label(query_cancer, text=\"Enter RefSeq Accession Number:\",width=55, height=2, font=(\"Copperplate\", 10, \"bold\"), fg=\"white\", bg=\"black\")\n",
    "            refrence_label.place(x=1160, y=400) \n",
    "\n",
    "            button_c = tk.Button(query_cancer, text=\"Retrieve clinical \\n data\", command=self.retrieve_clinical_data,width=20, height=2, font=(\"Copperplate\", 15, \"bold\"), fg=\"#F6B195\", bg=\"black\")\n",
    "            button_c.place(x=1160, y=500)\n",
    "\n",
    "            self.refrence_entry = tk.Entry(query_cancer)\n",
    "            self.refrence_entry.place(x=1160, y=450)\n",
    "\n",
    "        #####################################################        \n",
    "\n",
    "            # seperator box for deleting data\n",
    "            sep_del = tk.Label(query_cancer, text=\" Delete the record using ID  :\",width=55, height=2, font=(\"Copperplate\", 10, \"bold\"), fg=\"#F6B195\", bg=\"black\",anchor='w')\n",
    "            sep_del.place(x=768, y=40)\n",
    "\n",
    "            label_delete_id = tk.Label(query_cancer, text=\" Enter ID to delete:\",width=25, height=2, font=(\"Copperplate\", 10, \"bold\"), fg=\"white\", bg=\"black\")\n",
    "            label_delete_id.place(x=800, y=100, width=200)\n",
    "            \n",
    "            self.entry_delete_id = tk.Entry(query_cancer)\n",
    "            self.entry_delete_id.place(x=800, y=150)\n",
    "            \n",
    "            #retrieve Button:\n",
    "            button_retrieve = tk.Button(query_cancer, text=\"Retrieve DB\", command=self.retrieve_data,width=20, height=2, font=(\"Copperplate\", 20, \"bold\"), fg=\"red\", bg=\"white\")\n",
    "            button_retrieve.place(x=1600, y=1000)\n",
    "            \n",
    "           #count Button:\n",
    "            db_name_label = tk.Label(query_cancer, text=\"Database Name:\")\n",
    "            collection_name_label = tk.Label(query_cancer, text=\"Collection Name:\")\n",
    "\n",
    "            self.db_name_entry = tk.Entry(query_cancer)\n",
    "            self.collection_name_entry = tk.Entry(query_cancer)\n",
    "\n",
    "    #get count\n",
    "            sep_count = tk.Label(query_cancer, text=\"get the total count of records \",width=54, height=2,font=(\"Copperplate\", 10, \"bold\"), fg=\"#F6B195\", bg=\"black\")\n",
    "            sep_count.place(x=0, y=750)\n",
    "\n",
    "            get_count_button = tk.Button(query_cancer, text=\"Count records \", command=self.get_document_count,width=15, height=2, font=(\"Copperplate\", 15, \"bold\"), fg=\"#F6B195\", bg=\"black\")\n",
    "            get_count_button.place(x=50, y=800)\n",
    "\n",
    "            self.count_label = tk.Label(query_cancer, text=\"\",width=26, height=2,font=(\"Copperplate\", 15, \"bold\"), fg=\"#355C7D\", bg=\"#F6B195\",anchor='w')\n",
    "            self.count_label.place(x=50, y=890) \n",
    " #---------------#retieve one :---------------------------------\n",
    "            #Find one doc func\n",
    "    \n",
    "            sep_up_re= tk.Label(query_cancer, text=\" Retrieve One document: \",width=54, height=2,font=(\"Copperplate\", 10, \"bold\"), fg=\"#F6B195\", bg=\"black\")\n",
    "            sep_up_re.place(x=1165, y=610)\n",
    "    \n",
    "            button = tk.Button(query_cancer, text=\"Retrieve\", command=self.retrieve_document,width=10, height=2, font=(\"Copperplate\", 25, \"bold\"), fg=\"#F6B195\", bg=\"black\")\n",
    "            button.place(  x=1180, y=820)          \n",
    "    \n",
    "            label_one_doc = tk.Label(query_cancer, text=\" RefSeq Accession Number:\",width=25, height=2, font=(\"Copperplate\", 10, \"bold\"), fg=\"white\", bg=\"black\")\n",
    "            label_one_doc.place(x=1180, y=650)\n",
    "\n",
    "            self.label_entry_doc = tk.Entry(query_cancer)\n",
    "            self.label_entry_doc.place(x=1180, y=700)\n",
    "\n",
    " #------------ PROJECTION\n",
    "            self.col1_label = tk.Label(query_cancer, text=\"Column 1:\",width=15, height=2, font=(\"Copperplate\", 10, \"bold\"), fg=\"white\", bg=\"black\")\n",
    "            self.col1_label.place(x=385, y=560)\n",
    "\n",
    "            self.col1_entry = tk.Entry(query_cancer)\n",
    "            self.col1_entry.place(x=600, y=560)\n",
    "\n",
    "            self.col2_label = tk.Label(query_cancer, text=\"Column 2:\",width=15, height=2, font=(\"Copperplate\", 10, \"bold\"), fg=\"white\", bg=\"black\")\n",
    "            self.col2_label.place(x=385, y=600)\n",
    "\n",
    "            self.col2_entry = tk.Entry(query_cancer)\n",
    "            self.col2_entry.place(x=600, y=600)\n",
    "\n",
    "            # Create button to trigger projection\n",
    "            self.projection_button = tk.Button(query_cancer, text=\"Projection\", command=self.projection, width=20, height=2, font=(\"Copperplate\", 15, \"bold\"), fg=\"#F6B195\", bg=\"black\",anchor='w')\n",
    "            self.projection_button.place(x=385, y=650)    \n",
    "#--------------------------------------------------------------------\n",
    "            # update cancer\n",
    "         \n",
    "            self.db_name_entry = tk.Entry(query_cancer)\n",
    "            collection_name_entry = tk.Entry(query_cancer)\n",
    "            accession_number_entry = tk.Entry(query_cancer)\n",
    "            gene_name_entry = tk.Entry(query_cancer)\n",
    "            \n",
    "            sep_up = tk.Label(query_cancer, text=\" Update gene name: \",width=54, height=2,font=(\"Copperplate\", 10, \"bold\"), fg=\"#F6B195\", bg=\"black\")\n",
    "            sep_up.place(x=768, y=350)\n",
    "            \n",
    "            label_name = tk.Label(query_cancer, text=\"DB name\", width=10, height=2, font=(\"Copperplate\", 10, \"bold\"), fg=\"white\", bg=\"black\")\n",
    "            label_name.place(x=800, y=400)\n",
    "            self.entry_namedb = tk.Entry(query_cancer)\n",
    "            self.entry_namedb.place(x=1000, y=400)\n",
    "\n",
    "            label_collection = tk.Label(query_cancer, text=\"collection_name\", width=15, height=2, font=(\"Copperplate\", 10, \"bold\"), fg=\"white\", bg=\"black\")\n",
    "            label_collection.place(x=800, y=440)\n",
    "            self.entry_collection = tk.Entry(query_cancer)\n",
    "            self.entry_collection.place(x=1000, y=440)\n",
    "\n",
    "            label_acc = tk.Label(query_cancer, text=\"accesion number \", width=15, height=2, font=(\"Copperplate\", 10, \"bold\"), fg=\"white\", bg=\"black\")\n",
    "            label_acc.place(x=800, y=480)\n",
    "            self.accession_number_entry = tk.Entry(query_cancer)\n",
    "            self.accession_number_entry.place(x=1000, y=480)\n",
    "\n",
    "            label_gene_name = tk.Label(query_cancer, text=\"gene name\", width=10, height=2, font=(\"Copperplate\", 10, \"bold\"), fg=\"white\", bg=\"black\")\n",
    "            label_gene_name.place(x=800, y=520)\n",
    "            self.gene_name_entry = tk.Entry(query_cancer)\n",
    "            self.gene_name_entry.place(x=1000, y=523)\n",
    "\n",
    "            query_cancer.mainloop()\n",
    "\n",
    "#$$$end mongo \n",
    "            \n",
    "        #buttons\n",
    "        button1 = tk.Button(chcollection, text=\"Make queries on < Cancer data collection >\", font=(\"Copperplate\", 25, \"bold\"), fg=\"#F6B195\", bg=\"black\", command=open_query_cancer)\n",
    "        button1.place(relx=0.5, rely=0.5, anchor=\"center\")\n",
    "\n",
    "        #$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$\n",
    "\n",
    "\n",
    "        def open_query_clinical():\n",
    "            query_clinical = tk.Toplevel()\n",
    "            query_clinical.geometry(\"600x400\")\n",
    "            query_clinical.configure(bg=\"gray\")\n",
    "            query_clinical.attributes(\"-fullscreen\", True)\n",
    "            screen_width = query_clinical.winfo_screenwidth()\n",
    "            screen_height = query_clinical.winfo_screenheight()\n",
    "\n",
    "            canvas = tk.Canvas(query_clinical, width=screen_width, height=screen_height)\n",
    "            canvas.place(x=0, y=0)\n",
    "\n",
    "            gradient_colors = [\"#F8B195\", \"#F67280\", \"#C06C84\", \"#6C5B7B\", \"#355C7D\"]\n",
    "            for i in range(len(gradient_colors)):\n",
    "                color = gradient_colors[i]\n",
    "                x0 = int((i / len(gradient_colors)) * screen_width)\n",
    "                x1 = int(((i + 1) / len(gradient_colors)) * screen_width)\n",
    "                canvas.create_rectangle(x0, 0, x1, screen_height, fill=color, width=0)\n",
    "            #buttonbb = tk.Button(query_clinical, text=\"Back\", font=(\"Copperplate\", 25, \"bold\"), fg=\"#F6B195\", bg=\"black\", command=query_clinical.destroy)\n",
    "            #buttonbb.place(x=1500, y=900, width=200, height=70)\n",
    "            \n",
    "            buttonbb = tk.Button(query_clinical, text=\"Back\", font=(\"Copperplate\", 10, \"bold\"), fg=\"#F6B195\", bg=\"black\", command=query_clinical.destroy)\n",
    "            buttonbb.place(x=0,y=0, width=100, height=40)\n",
    "            \n",
    "            label2 = tk.Label(query_clinical, text=\"Label 2\", anchor=\"center\", cursor=\"star\")\n",
    "            label2.place(x=100, y=0, width=1900, height=40)\n",
    "            label2.config(text=\"Big data clinical file queries \", anchor=\"center\", cursor=\"star\", bg=\"black\",\n",
    "                              font=(\"Copperplate\", 15, \"bold\"), fg=\"#F6B195\")\n",
    "            \n",
    "######################################33#Update Many\n",
    "             # Create a label for the field input\n",
    "            field_label = tk.Label(query_clinical, text=\"Field:\",width=15, height=2, font=(\"Copperplate\", 10, \"bold\"), fg=\"white\", bg=\"black\")\n",
    "            field_label.place(x=1580, y=100)\n",
    "\n",
    "            # Create a textbox for field input\n",
    "            self.field_entry = tk.Entry(query_clinical)\n",
    "            self.field_entry.place(x=1580, y=130)\n",
    "\n",
    "            # Create a label for the old value input\n",
    "            old_value_label = tk.Label(query_clinical, text=\"Old Value:\",width=15, height=2, font=(\"Copperplate\", 10, \"bold\"), fg=\"white\", bg=\"black\")\n",
    "            old_value_label.place(x=1580, y=160)\n",
    "\n",
    "            # Create a textbox for old value input\n",
    "            self.old_value_entry = tk.Entry(query_clinical)\n",
    "            self.old_value_entry.place(x=1580, y=190)\n",
    "\n",
    "            # Create a label for the new value input\n",
    "            new_value_label = tk.Label(query_clinical, text=\"New Value:\",width=15, height=2, font=(\"Copperplate\", 10, \"bold\"), fg=\"white\", bg=\"black\")\n",
    "            new_value_label.place(x=1580, y=220)\n",
    "\n",
    "            # Create a textbox for new value input\n",
    "            self.new_value_entry = tk.Entry(query_clinical)\n",
    "            self.new_value_entry.place(x=1580, y=250)\n",
    "\n",
    "            # Create a button to trigger the update operation\n",
    "            update_button = tk.Button(query_clinical, text=\"Update Many Documents\", command=self.update_Many_documents, width=20, height=2, font=(\"Copperplate\", 15, \"bold\"), fg=\"#F6B195\", bg=\"black\",anchor='w')\n",
    "            update_button.place(x=1580, y=300)\n",
    "            ##########################################33\n",
    "            \n",
    " #---------------#retieve one :---------------------------------\n",
    "            #Find one doc func\n",
    "    \n",
    "            sep_up_re= tk.Label(query_clinical, text=\" Retrieve One document: \",width=54, height=2,font=(\"Copperplate\", 10, \"bold\"), fg=\"#F6B195\", bg=\"black\")\n",
    "            sep_up_re.place(x=1600, y=610)\n",
    "    \n",
    "            button = tk.Button(query_clinical, text=\"Retrieve\", command=self.retrieve_document_cli,width=10, height=2, font=(\"Copperplate\", 25, \"bold\"), fg=\"#F6B195\", bg=\"black\")\n",
    "            button.place(  x=1600, y=820)          \n",
    "    \n",
    "            label_one_doc = tk.Label(query_clinical, text=\" RefSeq Accession Number:\",width=25, height=2, font=(\"Copperplate\", 10, \"bold\"), fg=\"white\", bg=\"black\")\n",
    "            label_one_doc.place(x=1600, y=650)\n",
    "\n",
    "            self.label_entry_doc = tk.Entry(query_clinical)\n",
    "            self.label_entry_doc.place(x=1600, y=780)\n",
    "################################3\n",
    "\n",
    "          #---insert  \n",
    "            label_id = tk.Label(query_clinical, text=\" ACC:\", width=10, height=2, font=(\"Copperplate\", 10, \"bold\"), fg=\"white\", bg=\"black\",anchor='w')\n",
    "            label_id.place(x=30, y=60)\n",
    "            self.entry_accnum = tk.Entry(query_clinical)\n",
    "            self.entry_accnum.place(x=230, y=60)\n",
    "        \n",
    "            label_id = tk.Label(query_clinical, text=\" ID:\", width=10, height=2, font=(\"Copperplate\", 10, \"bold\"), fg=\"white\", bg=\"black\",anchor='w')\n",
    "            label_id.place(x=30, y=80)\n",
    "            self.entry_ID = tk.Entry(query_clinical)\n",
    "            self.entry_ID.place(x=230, y=90) \n",
    "            \n",
    "\n",
    "            label_symbol = tk.Label(query_clinical, text=\" Geneder :\",width=14, height=2, font=(\"Copperplate\", 10, \"bold\"), fg=\"white\", bg=\"black\",anchor='w')\n",
    "            label_symbol.place(x=30, y=113)\n",
    "            self.entry_Gender = tk.Entry(query_clinical)\n",
    "            self.entry_Gender.place(x=230, y=113)\n",
    "\n",
    "            label_name = tk.Label(query_clinical, text=\" ER status :\",width=14, height=2, font=(\"Copperplate\", 10, \"bold\"), fg=\"white\", bg=\"black\",anchor='w')\n",
    "            label_name.place(x=30, y=140)\n",
    "            self.entry_ER = tk.Entry(query_clinical)\n",
    "            self.entry_ER.place(x=230, y=140)\n",
    "\n",
    "            label_s1 = tk.Label(query_clinical, text=\" PR status:\",width=14, height=2, font=(\"Copperplate\", 10, \"bold\"), fg=\"white\", bg=\"black\",anchor='w')\n",
    "            label_s1.place(x=30, y=170)\n",
    "            self.entry_PR = tk.Entry(query_clinical)\n",
    "            self.entry_PR.place(x=230, y=170)\n",
    "            \n",
    "            label_s2 = tk.Label(query_clinical, text=\" Node:\",width=15, height=2, font=(\"Copperplate\", 10, \"bold\"), fg=\"white\", bg=\"black\",anchor='w')\n",
    "            label_s2.place(x=30, y=210)\n",
    "            self.entry_Node = tk.Entry(query_clinical)\n",
    "            self.entry_Node.place(x=230, y=210 )\n",
    "\n",
    "            label_s2 = tk.Label(query_clinical, text=\" Node code:\",width=10, height=1, font=(\"Copperplate\", 10, \"bold\"), fg=\"white\", bg=\"black\",anchor='w')\n",
    "            label_s2.place(x=30, y=250)\n",
    "            self.entry_Node_Coded = tk.Entry(query_clinical)\n",
    "            self.entry_Node_Coded.place(x=230, y=250)\n",
    "\n",
    "            label_s3 = tk.Label(query_clinical, text=\" Metastasis status:\",width=20, height=2, font=(\"Copperplate\", 10, \"bold\"), fg=\"white\", bg=\"black\",anchor='w')\n",
    "            label_s3.place(x=30, y=280)\n",
    "            self.entry_Metastasis = tk.Entry(query_clinical)\n",
    "            self.entry_Metastasis.place(x=230, y=280)\n",
    "\n",
    "            label_s4 = tk.Label(query_clinical, text=\"Metastasis Code:\",width=20, height=2, font=(\"Copperplate\", 10, \"bold\"), fg=\"white\", bg=\"black\",anchor='w')\n",
    "            label_s4.place(x=30, y=325)\n",
    "            self.entry_Metastasis_Coded = tk.Entry(query_clinical)\n",
    "            self.entry_Metastasis_Coded.place(x=230, y=330)\n",
    "\n",
    "            label_s5 = tk.Label(query_clinical, text=\" Survival status: \",width=20, height=2, font=(\"Copperplate\", 10, \"bold\"), fg=\"white\", bg=\"black\",anchor='w')\n",
    "            label_s5.place(x=30, y=367)\n",
    "            self.entry_survive = tk.Entry(query_clinical)\n",
    "            self.entry_survive.place(x=230, y=370)\n",
    "\n",
    "            label_s6 = tk.Label(query_clinical, text=\" PAM50 mRNA score :\",width=20, height=2, font=(\"Copperplate\", 10, \"bold\"), fg=\"white\", bg=\"black\",anchor='w')\n",
    "            label_s6.place(x=30, y=405)\n",
    "            self.entry_PAM50_mRNA = tk.Entry(query_clinical)\n",
    "            self.entry_PAM50_mRNA.place(x=230, y=410)\n",
    "\n",
    "            label_s7 = tk.Label(query_clinical, text=\" Methylation cluster score:\",width=27, height=2, font=(\"Copperplate\", 10, \"bold\"), fg=\"white\", bg=\"black\",anchor='w')\n",
    "            label_s7.place(x=30, y=445)\n",
    "            self.entry_methylation_clust = tk.Entry(query_clinical)\n",
    "            self.entry_methylation_clust.place(x=230, y=450)\n",
    "\n",
    "            label_s8 = tk.Label(query_clinical, text=\" Gene_exp_cluster:\",width=25, height=2, font=(\"Copperplate\", 10, \"bold\"), fg=\"white\", bg=\"black\",anchor='w')\n",
    "            label_s8.place(x=30, y=485)\n",
    "            self.entry_gebe_exp_clust = tk.Entry(query_clinical)\n",
    "            self.entry_gebe_exp_clust.place(x=230, y=490)\n",
    "\n",
    "            label_s9 = tk.Label(query_clinical, text=\"Age at initial diagnosis:\",width=25, height=2, font=(\"Copperplate\", 10, \"bold\"), fg=\"white\", bg=\"black\",anchor='w')\n",
    "            label_s9.place(x=30, y=525)\n",
    "            self.entry_Age_InitDiag = tk.Entry(query_clinical)\n",
    "            self.entry_Age_InitDiag.place(x=230, y=530)\n",
    "            \n",
    "            label_s10 = tk.Label(query_clinical, text=\" HER2 Final Status:\",width=25, height=2, font=(\"Copperplate\", 10, \"bold\"), fg=\"white\", bg=\"black\",anchor='w')\n",
    "            label_s10.place(x=30, y=560)\n",
    "            self.entry_HER2_FinalStatus = tk.Entry(query_clinical)\n",
    "            self.entry_HER2_FinalStatus.place(x=230, y=570)\n",
    "\n",
    "            label_s10 = tk.Label(query_clinical, text=\" Survival code:\",width=15, height=2, font=(\"Copperplate\", 10, \"bold\"), fg=\"white\", bg=\"black\",anchor='w')\n",
    "            label_s10.place(x=30, y=593)\n",
    "            self.entry_survive_code = tk.Entry(query_clinical)\n",
    "            self.entry_survive_code.place(x=230, y=600)      \n",
    "            \n",
    "            label_s10 = tk.Label(query_clinical, text=\" OS time:\",width=15, height=2, font=(\"Copperplate\", 10, \"bold\"), fg=\"white\", bg=\"black\",anchor='w')\n",
    "            label_s10.place(x=30, y=650)\n",
    "            self.entry_os_time = tk.Entry(query_clinical)\n",
    "            self.entry_os_time.place(x=230, y=650)\n",
    "            \n",
    "\n",
    "            ##################################\n",
    "#Retrieve 2 col \n",
    "             # First column input\n",
    "            col1_label = tk.Label(query_clinical, text=\"First Column:\",width=15, height=2, font=(\"Copperplate\", 10, \"bold\"), fg=\"white\", bg=\"black\")\n",
    "            col1_label.place(x=1180, y=150)\n",
    "            self.col1_entry = tk.Entry(query_clinical)\n",
    "            self.col1_entry.place(x=1180, y=200)\n",
    "\n",
    "            # Second column input\n",
    "            col2_label = tk.Label(query_clinical, text=\"Second Column:\",width=15, height=2, font=(\"Copperplate\", 10, \"bold\"), fg=\"white\", bg=\"black\")\n",
    "            col2_label.place(x=1180, y=250)\n",
    "            self.col2_entry = tk.Entry(query_clinical)\n",
    "            self.col2_entry.place(x=1180, y=300)\n",
    "\n",
    "            # Retrieve Columns button\n",
    "            button = tk.Button(query_clinical, text=\" Retrieve two Columns\", command=self.retrieve_columns, width=20, height=2, font=(\"Copperplate\", 15, \"bold\"), fg=\"#F6B195\", bg=\"black\",anchor='w')\n",
    "            button.place(x=1180, y=360)\n",
    "######################################\n",
    "\n",
    "#cli cli cli \n",
    "            # update clinical\n",
    "         \n",
    "            self.db_name_entry= tk.Entry(query_clinical)\n",
    "            collection_name_entry = tk.Entry(query_clinical)\n",
    "            ID_entry = tk.Entry(query_clinical)\n",
    "            survive_entry = tk.Entry(query_clinical)\n",
    "            \n",
    "            label_name = tk.Label(query_clinical, text=\" DB name\", width=10, height=2, font=(\"Copperplate\", 10, \"bold\"), fg=\"white\", bg=\"black\")\n",
    "            label_name.place(x=800, y=300)\n",
    "            \n",
    "            self.db_name_entry = tk.Entry(query_clinical)\n",
    "            self.db_name_entry.place(x=910, y=300)\n",
    "\n",
    "            \n",
    "            label_collection = tk.Label(query_clinical, text=\" collection name\", width=15, height=2, font=(\"Copperplate\", 10, \"bold\"), fg=\"white\", bg=\"black\")\n",
    "            label_collection.place(x=800, y=340)\n",
    "            \n",
    "            self.collection_name_entry= tk.Entry(query_clinical)\n",
    "            self.collection_name_entry.place(x=910, y=340)\n",
    "\n",
    "            label_acc = tk.Label(query_clinical, text=\" Enter ID \", width=15, height=2, font=(\"Copperplate\", 10, \"bold\"), fg=\"white\", bg=\"black\")\n",
    "            label_acc.place(x=800, y=380)\n",
    "            self.ID_entry = tk.Entry(query_clinical)\n",
    "            self.ID_entry.place(x=910, y=380)\n",
    "\n",
    "            label_gene_name = tk.Label(query_clinical, text=\" Enter patient \\nstatus\", width=15, height=2, font=(\"Copperplate\", 10, \"bold\"), fg=\"white\", bg=\"black\")\n",
    "            label_gene_name.place(x=800, y=420)\n",
    "            self.survive_entry = tk.Entry(query_clinical)\n",
    "            self.survive_entry.place(x=910, y=420)\n",
    " #------------ensure INDEXING IN AGE INIT\n",
    "            # Add a label and entry field for the field name\n",
    "            label = tk.Label(query_clinical, text=\" Enter Field Name:\",width=20, height=2, font=(\"Copperplate\", 10, \"bold\"), fg=\"white\", bg=\"black\")\n",
    "            label.place(x=800, y=100)\n",
    "            self.entry = tk.Entry(query_clinical)\n",
    "            self.entry.place(x=800, y=120)\n",
    "\n",
    "            # Add a button for creating the index\n",
    "            index_button = tk.Button(query_clinical, text=\"Create Index\", command=self.ensure_index, width=20, height=2, font=(\"Copperplate\", 15, \"bold\"), fg=\"#F6B195\", bg=\"black\",anchor='w')\n",
    "            index_button.place(x=800, y=180)\n",
    "            # Add a label for displaying the result\n",
    "            self.result_label = tk.Label(query_clinical, text=\"\")\n",
    "            self.result_label.place(x=800, y=160)\n",
    "                        # Create a button to get the indexes\n",
    "           \n",
    " #--------------buttons on screen : \n",
    "            # Create a button to trigger the $sum patients\n",
    "            calculate_button = tk.Button(query_clinical, text=\"Calculate Sum of \\nDECEASED patients\", command=self.calculate_sum,width=15, height=2, font=(\"Copperplate\", 15, \"bold\"), fg=\"#F6B195\", bg=\"black\",anchor='w')\n",
    "            calculate_button.place(x=400,y=930)\n",
    "\n",
    "            # Create a button to display records as CSV\n",
    "\n",
    "            insert_button = tk.Button(query_clinical, text=\"Insert\", command=self.insert_data_cli, width=10, height=2, font=(\"Copperplate\", 25, \"bold\"), fg=\"#F6B195\", bg=\"black\",anchor='w')\n",
    "            insert_button.place(x=50, y=710)\n",
    "\n",
    "            update_button = tk.Button(query_clinical, text=\"Update\",command=self.update_document_cli, width=10, height=2, font=(\"Copperplate\", 25, \"bold\"), fg=\"#F6B195\", bg=\"black\",anchor='w')\n",
    "            update_button.place(x=800, y=500)\n",
    "\n",
    "           #match males \n",
    "            button_find_males = tk.Button(query_clinical, text=\"Find Males\", command=self.find_males, width=20, height=2, font=(\"Copperplate\", 15, \"bold\"), fg=\"#F6B195\", bg=\"black\",anchor='w')\n",
    "            button_find_males.place(x=400, y=350)\n",
    "\n",
    " #timmmmme           #calculate_AVG_button = tk.Button(query_clinical, text=\"Calculate Average\", command=self.calculate_avg, width=20, height=2, font=(\"Copperplate\", 15, \"bold\"), fg=\"#F6B195\", bg=\"black\",anchor='w'  )\n",
    "            #calculate_AVG_button.place(x=400, y=450)\n",
    "            #retrieve Button:\n",
    "            button_retrieve = tk.Button(query_clinical, text=\"Retrieve DB\", command=self.retrieve_data_cli,width=20, height=2, font=(\"Copperplate\", 20, \"bold\"), fg=\"red\", bg=\"white\")\n",
    "            button_retrieve.place(x=1600, y=1000)\n",
    "            \n",
    "            # Create a button to show the min\n",
    "            \n",
    "            button = tk.Button(query_clinical, text=\"Show min age Result\", command=self.show_min, width=20, height=2, font=(\"Copperplate\", 15, \"bold\"), fg=\"#F6B195\", bg=\"black\",anchor='w')\n",
    "            button.place(x=400, y=150)\n",
    "            # Create a button to show the min\n",
    "\n",
    "            button = tk.Button(query_clinical, text=\"Show max age Results\", command=self.show_max, width=20, height=2, font=(\"Copperplate\", 15, \"bold\"), fg=\"#F6B195\", bg=\"black\",anchor='w')\n",
    "            button.place(x=400, y=250)\n",
    "            \n",
    "            # Create the \"min&limit\" button\n",
    "\n",
    "            # Create a button to trigger the search\n",
    "            search_button = tk.Button(query_clinical, text=\"Get all ages of patients \\n not equal <68> \", command=self.search_records,width=25, height=2, font=(\"Copperplate\", 15, \"bold\"), fg=\"#F6B195\", bg=\"black\",anchor='w')\n",
    "            search_button.place(x=400,y=850)\n",
    "\n",
    "            # &or doocs\n",
    "            # Create labels and text boxes for input\n",
    "            field1_label = tk.Label(query_clinical, text=\"Field 1:\",width=15, height=2, font=(\"Copperplate\", 10, \"bold\"), fg=\"white\", bg=\"black\")\n",
    "            field1_label.place(x=800, y=600)\n",
    "            self.field1_entry = tk.Entry(query_clinical)\n",
    "            self.field1_entry.place(x=920, y=600)\n",
    "\n",
    "            value1_label = tk.Label(query_clinical, text=\"Value 1:\",width=15, height=2, font=(\"Copperplate\", 10, \"bold\"), fg=\"white\", bg=\"black\")\n",
    "            value1_label.place(x=800, y=640)\n",
    "            self.value1_entry = tk.Entry(query_clinical)\n",
    "            self.value1_entry.place(x=920, y=640)\n",
    "\n",
    "            field2_label = tk.Label(query_clinical, text=\"Field 2:\",width=15, height=2, font=(\"Copperplate\", 10, \"bold\"), fg=\"white\", bg=\"black\")\n",
    "            field2_label.place(x=800, y=700)\n",
    "            self.field2_entry = tk.Entry(query_clinical)\n",
    "            self.field2_entry.place(x=920, y=700)\n",
    "\n",
    "            value2_label = tk.Label(query_clinical, text=\"Value 2:\",width=15, height=2, font=(\"Copperplate\", 10, \"bold\"), fg=\"white\", bg=\"black\")\n",
    "            value2_label.place(x=800, y=740)\n",
    "            60\n",
    "            self.value2_entry = tk.Entry(query_clinical)\n",
    "            self.value2_entry.place(x=920, y=740)\n",
    "\n",
    "            # Create the retrieve button << Gender >>\n",
    "            retrieve_button = tk.Button(query_clinical, text=\"$or\", command=self.retrieve_or_documents, width=20, height=2, font=(\"Copperplate\", 15, \"bold\"), fg=\"#F6B195\", bg=\"black\",anchor='w')\n",
    "            retrieve_button.place(x=800, y=800)\n",
    "            # Create the \"max &limit\" button\n",
    "            \n",
    "\n",
    "            button_sort= tk.Button(query_clinical, text=\"Display Records of \\n sorted <dec> age females\", command=self.display_records,font=(\"Copperplate\", 15, \"bold\"), fg=\"#F6B195\", bg=\"black\")\n",
    "            button_sort.place(x=400, y=45, width=350, height=80)\n",
    "            \n",
    "            button_sort= tk.Button(query_clinical, text=\"Display Records of \\n sorted <asc> age females\", command=self.display_records_one,font=(\"Copperplate\", 15, \"bold\"), fg=\"#F6B195\", bg=\"black\")\n",
    "            button_sort.place(x=400, y=750, width=350, height=80)\n",
    "\n",
    "            \n",
    "            #LIMIT ON CLINICAL DATA\n",
    "            label = tk.Label(query_clinical, text=\"Enter the number of \\n documents to retrieve:\",width=25, height=2, font=(\"Copperplate\", 10, \"bold\"), fg=\"white\", bg=\"black\")\n",
    "            label.place(x=1180, y=500)\n",
    "\n",
    "            self.limit_entry = tk.Entry(query_clinical)\n",
    "            self.limit_entry.place(x=1180, y=550)\n",
    "\n",
    "            button = tk.Button(query_clinical, text=\"Retrieve \\n Limited Documents\", command=self.limit_documents, width=20, height=2, font=(\"Copperplate\", 15, \"bold\"), fg=\"#F6B195\", bg=\"black\",anchor='w')\n",
    "            button.place(x=1180, y=570)\n",
    "          ################################################\n",
    "\n",
    "\n",
    "            query_clinical.mainloop()\n",
    "\n",
    "            \n",
    "        #buttons\n",
    "        buttonp = tk.Button(chcollection, text=\"Make queries on <Clinical data collection>\", font=(\"Copperplate\", 25, \"bold\"), fg=\"#F6B195\", bg=\"black\", command=open_query_clinical )\n",
    "        buttonp.place(relx=0.5, rely=0.6, anchor=\"center\")\n",
    "        \n",
    "\n",
    "############ page2: start page\n",
    "\n",
    "    def show_hello(self):\n",
    "        self.hello.deiconify()  # Show the \"Hello\" window\n",
    "        self.chcollection.destroy()  # Destroy the \"chcollection\" window if it exists\n",
    "        self.hello_update.deiconify() \n",
    "\n",
    "    def startpage(self):\n",
    "        if self.hello:\n",
    "\n",
    "            self.hello.deiconify()  # Show the \"Big data\" window if it exists\n",
    "        else:\n",
    "            self.hello = tk.Tk()  # Create the \"Big data\" window\n",
    "            self.hello.title(\"start page\")\n",
    "            self.hello.geometry(\"200x100\")\n",
    "            # self.hello.resizable(width=False, height=False)\n",
    "            self.hello.configure(bg=\"gray\")\n",
    "            # Configure the window to cover the entire screen\n",
    "            screen_width = self.hello.winfo_screenwidth()\n",
    "            screen_height = self.hello.winfo_screenheight()\n",
    "            self.hello.geometry(f\"{screen_width}x{screen_height}+0+0\")\n",
    "\n",
    "            # Create a canvas and place it at coordinates (0, 0) to cover the entire window\n",
    "            canvas = tk.Canvas(self.hello, width=screen_width, height=screen_height)\n",
    "            canvas.place(x=0, y=0)\n",
    "\n",
    "            # Define the pastel gradient colors\n",
    "            gradient_colors = [\"#F8B195\", \"#F67280\", \"#C06C84\", \"#6C5B7B\", \"#355C7D\"]\n",
    "\n",
    "            # Labels:\n",
    "            label1 = tk.Label(self.hello, text=\"Label 1\", anchor=\"center\", cursor=\"star\", bg=\"black\",\n",
    "                              font=(\"Copperplate\", 35, \"bold\"), fg=\"#F6B195\")\n",
    "            label1.place(x=0, y=20, width=1900, height=300)\n",
    "\n",
    "            label1.config(text=\"Welcome to Breast Cancer Proteomes big data \\n queries using mongoBD\")\n",
    "\n",
    "            #---screen buttons:\n",
    "            def show_label1():\n",
    "                if label2.cget(\"text\") == \"\":\n",
    "                    def show_databases(self):\n",
    "                        # Connect to MongoDB\n",
    "                        client = MongoClient('mongodb://localhost:27017/')\n",
    "\n",
    "                        # Get the list of database names\n",
    "                        database_names = client.list_database_names()\n",
    "\n",
    "                        # Display the list of databases\n",
    "                        #messagebox.showinfo(\"MongoDB Databases\", \"\\n\".join(database_names),anchor=\"center\", cursor=\"star\", bg=\"black\", font=(\"Copperplate\", 35, \"bold\"), fg=\"#F6B195\")\n",
    "                        label3.config(text=\"MongoDB Databases:\\n\" + \"\\n\".join(database_names), anchor=\"w\", cursor=\"star\", bg=\"black\", font=(\"Copperplate\", 35, \"bold\"), fg=\"#F6B195\")\n",
    "                        label3.place(x=400, y=600)\n",
    "\n",
    "\n",
    "                else:\n",
    "                    label2.config(text=\"\", bg=\"white\")\n",
    "\n",
    "            def show_label2():\n",
    "                if label3.cget(\"text\") == \"\":\n",
    "                    label3.config(text=\"You have two Collections:\\n 1- Cancer data \\n 1- Clinical_data \", anchor=\"w\", cursor=\"star\", bg=\"black\", font=(\"Copperplate\", 35, \"bold\"), fg=\"#F6B195\")\n",
    "                    label3.place(x=400, y=600)\n",
    "                else:\n",
    "                    label3.config(text=\"\", bg=\"white\")\n",
    "                    \n",
    "            button1 = tk.Button(self.hello, text=\"Show My Databases\", command=self.show_databases, font=(\"Copperplate\", 25, \"bold\"), fg=\"#F6B195\", bg=\"black\")\n",
    "            button1.place(x=0, y=410, width=400, height=100)\n",
    "            \n",
    "\n",
    "\n",
    "            button2 = tk.Button(self.hello, text=\"Show Collections\", font=(\"Copperplate\", 25, \"bold\"), fg=\"#F6B195\", bg=\"black\", command=show_label2)\n",
    "            button2.place(x=0, y=600, width=400, height=100)\n",
    "\n",
    "            label2 = tk.Label(self.hello, text=\"\", font=(\"Arial\", 12), bg=\"white\")\n",
    "            label3 = tk.Label(self.hello, text=\"\", font=(\"Arial\", 12), bg=\"white\")\n",
    "            \n",
    "            button3 = tk.Button(self.hello, text=\"Start\", font=(\"Copperplate\", 25, \"bold\"), fg=\"#F6B195\", bg=\"black\", command=self.open_chcollection)\n",
    "            button3.place(x=0, y=90, width=200, height=70)\n",
    "           \n",
    "\n",
    "            outbutton = tk.Button(self.hello, text=\"Exit\", width=10, height=2, font=(\"Copperplate\", 25, \"bold\"), fg=\"#F6B195\", bg=\"black\", command=self.hello.destroy)\n",
    "            outbutton.place(x=0, y=200, width=200, height=70)\n",
    "\n",
    "            # Create a rectangle with a gradient background that covers the entire canvas\n",
    "            for i in range(len(gradient_colors) - 1):\n",
    "                y_start = int((i / (len(gradient_colors) - 1)) * screen_height)\n",
    "                y_end = int(((i + 1) / (len(gradient_colors) - 1)) * screen_height)\n",
    "                canvas.create_rectangle(0, y_start, screen_width, y_end, fill=gradient_colors[i], width=0)\n",
    "\n",
    "        self.retrieve_data()  # Call the retrieve_data method to display the data table\n",
    "\n",
    "        self.hello.mainloop()\n",
    "#FUNCTION 1  %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%INSERT%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%\n",
    "\n",
    "    def insert_data(self):\n",
    "        # Check if the ID already exists in the collection\n",
    "        existing_document = self.collection.find_one({'RefSeq_accession_number': self.entry_id.get()})\n",
    "        if existing_document:\n",
    "            # Show a message box indicating that the ID already exists\n",
    "            messagebox.showinfo('Error', 'ID already exists. Cannot insert the document.')\n",
    "        else:\n",
    "            # Create a document to insert\n",
    "            data = {\n",
    "                'RefSeq_accession_number': self.entry_id.get(),\n",
    "                'gene_symbol': self.entry_symbol.get(),\n",
    "                'gene_name': self.entry_name.get(),\n",
    "                'AO_A12D': self.entry_s1.get(),\n",
    "                'C8_A131': self.entry_s2.get(),\n",
    "                'BH_A18Q': self.entry_s3.get(),\n",
    "                'C8_A130': self.entry_s4.get(),\n",
    "                'C8_A138': self.entry_s5.get(),\n",
    "                'E2_A154': self.entry_s6.get(),\n",
    "                'C8_A12L': self.entry_s7.get(),\n",
    "                'A2_A0EX': self.entry_s8.get(),\n",
    "                'A7_A0CE': self.entry_s9.get(),\n",
    "                'AN_A04A': self.entry_s10.get(),               \n",
    "\n",
    "            }\n",
    "\n",
    "            # Insert the document into the collection\n",
    "            self.collection.insert_one(data)\n",
    "            messagebox.showinfo('Success', 'Data inserted successfully.')\n",
    "\n",
    "#FUNCTION 2  %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%DELETE%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%\n",
    "\n",
    "    def delete_data(self):\n",
    "        # Check if the ID exists in the collection\n",
    "        existing_document = self.collection.find_one({'RefSeq_accession_number': self.entry_delete_id.get()})\n",
    "        if existing_document:\n",
    "            # Delete the document from the collection\n",
    "            self.collection.delete_one({'RefSeq_accession_number': self.entry_delete_id.get()})\n",
    "            messagebox.showinfo('Success', 'Data deleted successfully.')\n",
    "        else:\n",
    "            # Show a message box indicating that the ID does not exist\n",
    "            messagebox.showinfo('Error', 'ID does not exist. Cannot delete the document.')\n",
    "            \n",
    "#FUNCTION 3  %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%UPDATE%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%\n",
    "\n",
    "    def update_document(self):\n",
    "        client = MongoClient()\n",
    "\n",
    "    # Access the specified database and collection\n",
    "        db = client['breast']\n",
    "        collection = db['cancerf']\n",
    "            # Retrieve data from the collection\n",
    "        accession_number = self.accession_number_entry.get()\n",
    "        new_gene_name = self.gene_name_entry.get()\n",
    "        \n",
    "            # Update the document object based on the filter and update queries\n",
    "        filter_query = {'RefSeq_accession_number': accession_number}\n",
    "        update_query = {'$set': {'gene_name': new_gene_name}}\n",
    "        collection.update_one(filter_query, update_query)\n",
    "        messagebox.showinfo('sucess', 'gene name updated .')\n",
    "\n",
    "#FUNCTION 4  %%%%%%%%%%%%%%%%%%%%%%%%%%%%%% AVRAGE %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%\n",
    "                #Aggregate1\n",
    "\n",
    "    def calculate_avg(self):\n",
    "    # Define the aggregation pipeline\n",
    "        pipeline = [\n",
    "            {\n",
    "                '$group': {\n",
    "                    '_id': '$gene_name',\n",
    "                    'avg': {\n",
    "                        '$avg': {\n",
    "                            '$avg': ['$AO_A12D', '$C8_A131', '$BH_A18Q', '$C8_A130', '$C8_A138',\n",
    "                                     '$E2_A154', '$C8_A12L', '$A2_A0EX', '$A7_A0CE', '$AN_A04A',\n",
    "                                    ]\n",
    "                        }\n",
    "                    }\n",
    "                }\n",
    "            }\n",
    "        ]\n",
    "         \n",
    "\n",
    "        # Execute the aggregation pipeline\n",
    "        result_avg = list(self.collection.aggregate(pipeline))\n",
    "\n",
    "        # Create a new window to display the result\n",
    "        result_avg_window = tk.Toplevel(self.hello)\n",
    "        result_avg_window.title(\"Aggregation Result\")\n",
    "        #result_avg_window.attributes(\"-fullscreen\", True)\n",
    "\n",
    "        # Create a Treeview widget to display the result in a table-like structure\n",
    "        tree = ttk.Treeview(result_avg_window)\n",
    "        tree[\"columns\"] = (\"Gene Name\", \"Average\")\n",
    "        tree.heading(\"Gene Name\", text=\"Gene Name\")\n",
    "        tree.heading(\"Average\", text=\"Average\")\n",
    "\n",
    "        # Create a scrollbar\n",
    "        scrollbar = ttk.Scrollbar(result_avg_window, orient=\"vertical\", command=tree.yview)\n",
    "        scrollbar.pack(side=\"right\", fill=\"y\")\n",
    "\n",
    "        tree.configure(yscrollcommand=scrollbar.set)\n",
    "\n",
    "        # Insert the aggregation result into the Treeview widget\n",
    "        for doc in result_avg:\n",
    "            gene_name = doc['_id']\n",
    "            avg = doc['avg']\n",
    "            tree.insert(\"\", \"end\", values=(gene_name, avg))\n",
    "\n",
    "        # Pack the Treeview widget\n",
    "        tree.pack(fill=\"both\", expand=True)\n",
    "        result_avg_window.mainloop()\n",
    "#FUNCTION 5 %%%%%%%%%%%%%%%%%%%%%%%%%%%%%% get_document_count %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%\n",
    " \n",
    "    def get_document_count(self):\n",
    "        client = MongoClient()\n",
    "        # Access the specified database\n",
    "        db = client[\"breast\"]\n",
    "        # Access the specified collection\n",
    "        collection = db[\"cancerf\"]\n",
    "\n",
    "        db_name = self.db_name_entry.get()\n",
    "        collection_name = self.collection_name_entry.get()\n",
    "\n",
    "        # Get the count of all documents in the collection\n",
    "        count = collection.count_documents({})\n",
    "\n",
    "        # Display the count\n",
    "        self.count_label.configure(text=f\"The number of\\n records are: {count}\")\n",
    "        \n",
    "#FUNCTION ##########refrence to cancer data ########################################################################\n",
    "    def retrieve_cancer_data(self):\n",
    "            refseq_acc = self.entry.get()\n",
    "\n",
    "            # Connect to MongoDB\n",
    "            client = MongoClient('mongodb://localhost:27017/')\n",
    "            db = client['breast']  # Replace 'your_database_name' with your actual database name\n",
    "            clinical_data_collection = db['clinical_data']\n",
    "            cancer_collection = db['cancerf']\n",
    "\n",
    "            # Find the document in the 'clinical_data' collection with the specified RefSeq Accession Number\n",
    "            clinical_data_doc = clinical_data_collection.find_one({\"RefSeq_accession_number\": refseq_acc})\n",
    "\n",
    "            if clinical_data_doc:\n",
    "                # Get the RefSeq Accession Number from the 'clinical_data' collection document\n",
    "                refseq_acc_clinical = clinical_data_doc['RefSeq_accession_number']\n",
    "\n",
    "                # Find the document in the 'cancer' collection with the matching RefSeq Accession Number\n",
    "                cancer_doc = cancer_collection.find_one({\"RefSeq_accession_number\": refseq_acc_clinical})\n",
    "\n",
    "            if cancer_doc:\n",
    "                # Display the document in a message box\n",
    "                messagebox.showinfo(\"Document\", str(cancer_doc))\n",
    "            else:\n",
    "                messagebox.showinfo(\"Error\", \"Document not found.\")\n",
    "# FUNCT ######################### IN OPERATOR ##################3\n",
    "#################################################IN_operatot\n",
    "    def in_operator(self):\n",
    "        # Get the user input\n",
    "        accession1 = self.accession1_entry.get()\n",
    "        accession2 = self.accession2_entry.get()\n",
    "        accession3 = self.accession3_entry.get()\n",
    "\n",
    "        # Connect to MongoDB\n",
    "        client = MongoClient('mongodb://localhost:27017/')\n",
    "        db = client['breast']\n",
    "        collection = db['cancerf']\n",
    "\n",
    "        # Prepare the query using the $in operator\n",
    "        query = {\n",
    "            \"RefSeq_accession_number\": {\"$in\": [accession1, accession2, accession3]}\n",
    "        }\n",
    "\n",
    "        # Retrieve documents from the collection with the specified RefSeq_accession_numbers\n",
    "        documents = collection.find(query)\n",
    "\n",
    "        # Create a new window to display the result\n",
    "        result_window = tk.Toplevel()\n",
    "        result_window.title(\"Document Data\")\n",
    "        result_window.geometry(\"600x400\")\n",
    "\n",
    "        # Create a Treeview widget to display the result in a table-like structure\n",
    "        tree = ttk.Treeview(result_window)\n",
    "        tree[\"columns\"] = tuple(documents[0].keys()) if documents else ()\n",
    "\n",
    "        # Format the columns\n",
    "        for column in tree[\"columns\"]:\n",
    "            tree.column(column, anchor='center')\n",
    "            tree.heading(column, text=column)\n",
    "\n",
    "        # Insert the document data into the Treeview widget\n",
    "        for doc in documents:\n",
    "            tree.insert(\"\", \"end\", values=list(doc.values()))\n",
    "\n",
    "        # Add scrollbar\n",
    "        scrollbar = ttk.Scrollbar(result_window, orient=\"vertical\", command=tree.yview)\n",
    "        scrollbar.pack(side=\"right\", fill=\"y\")\n",
    "\n",
    "        tree.configure(yscrollcommand=scrollbar.set)\n",
    "\n",
    "        # Pack the Treeview widget\n",
    "        tree.pack(fill=\"both\", expand=True)\n",
    "#FUNCTION  6 %%%%%%%%%%%%%%%%%%%%%%%%%%%%%% get gene sym gt 1 %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%\n",
    "    def get_gene_symbols_with_greater_than_one(self):\n",
    "            # Connect to the MongoDB server\n",
    "            client = MongoClient()\n",
    "            db = client['breast']\n",
    "            collection = db['cancerf']\n",
    "\n",
    "            # Get the input values from the textboxes\n",
    "            self.db_name = self.db_name.get()\n",
    "            collection_name = self.collection_name.get()\n",
    "            object_name = self.object_name.get()\n",
    "\n",
    "            # Connect to the MongoDB server\n",
    "            client = MongoClient()\n",
    "\n",
    "            # Access the specified database and collection\n",
    "            db = client['breast']\n",
    "            collection = db['cancerf']\n",
    "\n",
    "            # Construct the filter query using the $expr operator\n",
    "            filter_query = {\"$expr\": { \"$gt\": [\"$\" + object_name, 1]}}\n",
    "\n",
    "            # Perform the find operation with projection\n",
    "            results = collection.find(filter_query, {\"gene_symbol\": 1})\n",
    "\n",
    "            # Extract the gene symbols from the results\n",
    "            gene_symbols = [result[\"gene_symbol\"] for result in results]\n",
    "\n",
    "            # Display the gene symbols in a messagebox\n",
    "            tk.messagebox.showinfo(\"Gene Symbols\", f\"The gene symbols are: {gene_symbols}\")\n",
    "\n",
    "\n",
    "#FUNCTION  8 %%%%%%%%%%%%%%%%%%%%%%%%%%%%%% gene name lt 1 %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%  \n",
    "    def get_gene_symbols_with_less_than_one(self):\n",
    "                # Connect to the MongoDB server\n",
    "            client = MongoClient()\n",
    "            db = client['breast']\n",
    "            collection = db['cancerf']\n",
    "\n",
    "            # Get the input values from the textboxes\n",
    "            self.db_name_less = self.db_name_entry_less.get()\n",
    "            collection_name_less = self.collection_name_entry_less.get()\n",
    "            object_name_less = self.object_name_entry_less.get()\n",
    "\n",
    "            client = MongoClient()\n",
    "\n",
    "            # Access the specified database and collection\n",
    "            db = client['breast']\n",
    "            collection = db['cancerf']\n",
    "\n",
    "            # Construct the filter query using the $expr operator\n",
    "            filter_querys = {\"$expr\": {\"$lt\": [\"$\" + object_name_less, 1]}}\n",
    "\n",
    "            # Perform the find operation with projection\n",
    "            resultss = collection.find(filter_querys, {\"gene_symbol\": 1})\n",
    "\n",
    "            # Extract the gene symbols from the results\n",
    "            gene_symbolss = [result[\"gene_symbol\"] for result in resultss]\n",
    "\n",
    "            \n",
    "\n",
    "            # Display the gene symbols in a messagebox\n",
    "            tk.messagebox.showinfo(\"Gene Symbols\", f\"The gene symbols are: {gene_symbolss}\")\n",
    "#FUNCTION  7 %%%%%%%%%%%%%%%%%%%%%%%%%%%%%% HASH %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%\n",
    "\n",
    "    def hash_document_id(self):\n",
    "        # Get the document ID from the entry field\n",
    "        document_id = self.hash_entry.get()\n",
    "\n",
    "        # Hash the document ID using SHA-256\n",
    "        hashed_id = hashlib.sha256(document_id.encode()).hexdigest()\n",
    "\n",
    "        # Connect to the MongoDB database\n",
    "        client = MongoClient('mongodb://localhost:27017')\n",
    "        db = client['breast']\n",
    "        collection = db['cancerf']\n",
    "\n",
    "        # Update the document with the hashed ID\n",
    "        collection.update_one({'RefSeq_accession_number': document_id}, {'$set': {'RefSeq_accession_number': '#' * len(hashed_id)}})\n",
    "        messagebox.showinfo(\"success\",\"ID Hashed Successfully !\")\n",
    " #####################Projection\n",
    "    def projection(self):\n",
    "        # Get the user input\n",
    "        col1 = self.col1_entry.get()\n",
    "        col2 = self.col2_entry.get()\n",
    "\n",
    "        # Connect to MongoDB\n",
    "        client = MongoClient('mongodb://localhost:27017/')\n",
    "        db = client['breast']\n",
    "        collection = db['cancerf']\n",
    "\n",
    "        # Prepare the projection fields\n",
    "        projection = {\n",
    "            col1: 1,\n",
    "            col2: 1,\n",
    "            '_id': 0\n",
    "        }\n",
    "\n",
    "        # Retrieve documents from the collection with projection\n",
    "        documents = collection.find({}, projection)\n",
    "\n",
    "        # Create a new window to display the result\n",
    "        result_window = tk.Toplevel()\n",
    "        result_window.title(\"Column Data\")\n",
    "        result_window.geometry(\"600x400\")\n",
    "\n",
    "        # Create a Treeview widget to display the result in a table-like structure\n",
    "        tree = ttk.Treeview(result_window)\n",
    "        tree[\"columns\"] = (col1, col2)\n",
    "        tree.heading(col1, text=col1)\n",
    "        tree.heading(col2, text=col2)\n",
    "\n",
    "        # Create a scrollbar\n",
    "        scrollbar = ttk.Scrollbar(result_window, orient=\"vertical\", command=tree.yview)\n",
    "        scrollbar.pack(side=\"right\", fill=\"y\")\n",
    "\n",
    "        tree.configure(yscrollcommand=scrollbar.set)\n",
    "\n",
    "        # Insert the column values into the Treeview widget\n",
    "        for document in documents:\n",
    "            values = [document.get(col1, \"\"), document.get(col2, \"\")]\n",
    "            tree.insert(\"\", \"end\", values=values)\n",
    "\n",
    "        # Pack the Treeview widget\n",
    "        tree.pack(fill=\"both\", expand=True)       \n",
    "#FUNCTION ############################# retrive one################################\n",
    "    def retrieve_document(self):\n",
    "            accession_number = self.label_entry_doc.get()\n",
    "\n",
    "            # Perform the query\n",
    "            document = self.collection.find_one({'RefSeq_accession_number': accession_number})\n",
    "\n",
    "            if document:\n",
    "                # Display the document in a message box\n",
    "                messagebox.showinfo(\"Document\", str(document))\n",
    "            else:\n",
    "                messagebox.showinfo(\"Error\", \"Document not found.\")\n",
    "                \n",
    "                \n",
    "#FUNCTION 2  %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%DELETEMANY%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%\n",
    "    def delete_Many_documents(self):\n",
    "        field = self.field_entry.get()\n",
    "        value = self.value_entry.get()\n",
    "        \n",
    "        # Check if any documents exist with the given field value\n",
    "        existing_documents = self.collection.count_documents({field: value})\n",
    "    \n",
    "        if existing_documents > 0:\n",
    "            # Delete all documents with the given field value\n",
    "            self.collection.delete_many({field: value})\n",
    "            messagebox.showinfo('Success', 'Documents deleted successfully.')\n",
    "        else:\n",
    "            # Show a message box indicating that no documents were found\n",
    "            messagebox.showinfo('Error', 'No documents found with the specified field value.')                \n",
    "#FUNCTION   %%%%%%%%%%%%%%%%%%%%%%%%%%%%%% retrieve %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%\n",
    "    def retrieve_data(self):\n",
    "        # Retrieve data from the collection\n",
    "        data = self.collection.find().sort('_id', DESCENDING)\n",
    "\n",
    "        # Create a new window for the results\n",
    "        results_window = tk.Toplevel(self.hello)\n",
    "        results_window.title(\"Results\")\n",
    "        results_window.geometry(\"800x600\")\n",
    "\n",
    "        # Create a frame to hold the table and scrollbars\n",
    "        frame = tk.Frame(results_window)\n",
    "        frame.pack(fill='both', expand=True)\n",
    "\n",
    "        # Create a vertical scrollbar\n",
    "        y_scrollbar = ttk.Scrollbar(frame, orient='vertical')\n",
    "        y_scrollbar.pack(side='right', fill='y')\n",
    "\n",
    "        # Create a horizontal scrollbar\n",
    "        x_scrollbar = ttk.Scrollbar(frame, orient='horizontal')\n",
    "        x_scrollbar.pack(side='bottom', fill='x')\n",
    "\n",
    "        # Create a treeview widget to display the data\n",
    "        treeview = ttk.Treeview(frame, yscrollcommand=y_scrollbar.set, xscrollcommand=x_scrollbar.set)\n",
    "        treeview.pack(fill='both', expand=True)\n",
    "\n",
    "        # Configure the scrollbars\n",
    "        y_scrollbar.configure(command=treeview.yview)\n",
    "        x_scrollbar.configure(command=treeview.xview)\n",
    "\n",
    "        # Define the columns\n",
    "        columns = tuple(data[0].keys())\n",
    "        treeview['columns'] = columns\n",
    "\n",
    "        # Configure the columns\n",
    "        for column in columns:\n",
    "            treeview.column(column, anchor='center', width=100)\n",
    "            treeview.heading(column, text=column)\n",
    "\n",
    "        # Add a border style to the treeview\n",
    "        style = ttk.Style()\n",
    "        style.configure(\"Treeview\", borderwidth=1, relief=\"solid\")\n",
    "\n",
    "        # Insert the data into the treeview\n",
    "        for document in data:\n",
    "            treeview.insert('', 'end', values=tuple(document.values()))\n",
    "\n",
    "        results_window.mainloop()\n",
    "        \n",
    "#---cli- FUNCTION  1 %%%%%%%%%%%%%%%%%%%%%%%%%%%%%% insert %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%\n",
    "    def insert_data_cli(self):\n",
    "        # Check if the ID already exists in the collection\n",
    "        existing_document = self.collection2.find_one({'RefSeq_accession_number': self.entry_accnum.get()})\n",
    "        if existing_document:\n",
    "            # Show a message box indicating that the ID already exists\n",
    "            messagebox.showinfo('Error', 'ACC already exists. Cannot insert the document.')\n",
    "        else:\n",
    "            # Create a document to insert\n",
    "            data = {\n",
    "\n",
    "                'RefSeq_accession_number': self.entry_accnum.get(), #\\\\\\\\\\\\\n",
    "                'ID': self.entry_ID.get(),\n",
    "                'Gender': self.entry_Gender.get(),\n",
    "                'ER': self.entry_ER.get(),\n",
    "                'PR': self.entry_PR.get(),\n",
    "                'Node': self.entry_Node.get(),\n",
    "                'Node_Coded': self.entry_Node_Coded.get(),\n",
    "                'Metastasis': self.entry_Metastasis.get(),\n",
    "                'Metastasis_Coded': self.entry_Metastasis_Coded.get(),\n",
    "                'survive': self.entry_survive.get(),\n",
    "                'PAM50_mRNA': self.entry_PAM50_mRNA.get(),\n",
    "                'methylation_clust': self.entry_methylation_clust.get(),\n",
    "                'gebe_exp_clust': self.entry_gebe_exp_clust.get(),\n",
    "                'Age_InitDiag': self.entry_Age_InitDiag.get(),               \n",
    "                'HER2_FinalStatus': self.entry_HER2_FinalStatus.get(),\n",
    "                'survive_code': self.entry_survive_code.get(),\n",
    "                'OS_Time': self.entry_os_time.get(),             \n",
    "\n",
    "            }\n",
    "\n",
    "            self.collection2.insert_one(data)\n",
    "            messagebox.showinfo('Success', 'Data inserted successfully.')\n",
    "            \n",
    "            \n",
    "    #---cli- FUNCTION  %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%  %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%\n",
    "    def find_males(self):\n",
    "        # Connect to MongoDB\n",
    "        client = MongoClient()\n",
    "\n",
    "        db = client['breast']\n",
    "        collection2 = db['clinical_data']\n",
    "\n",
    "        # Perform aggregation match query for males\n",
    "        pipeline = [\n",
    "            {\"$match\": {\"Gender\": \"MALE\"}}\n",
    "        ]\n",
    "        result = collection2.aggregate(pipeline)\n",
    "\n",
    "        # Create a new Tkinter window\n",
    "        result_window = tk.Toplevel()\n",
    "\n",
    "        # Iterate over the result and display the matched records in the window\n",
    "        for record in result:\n",
    "            for key, value in record.items():\n",
    "                label = tk.Label(result_window, text=f\"{key}: {value}\")\n",
    "                label.pack()\n",
    "###############INDEXING\n",
    "    def ensure_index(self):\n",
    "    # Get the field name from the entry field\n",
    "        field_name = self.entry.get()\n",
    "\n",
    "        # Connect to the MongoDB database\n",
    "        client = MongoClient('mongodb://localhost:27017')\n",
    "        db = client['breast']\n",
    "        collection = db['clinical_data']\n",
    "\n",
    "        # Create an index on the specified field\n",
    "        collection.create_index([(field_name, 1)])\n",
    "\n",
    "        # Check if the index exists\n",
    "        index_info = collection.index_information()\n",
    "        if field_name + \"_1\" in index_info:\n",
    "            self.result_label.config(text=\"Index added to {} field.\".format(field_name))\n",
    "        else:\n",
    "            self.result_label.config(text=\"Error: Failed to add index to {} field.\".format(field_name))\n",
    "\n",
    "#---cli- FUNCTION  2 %%%%%%%%%%%%%%%%%%%%%%%%%%%%%% sort female age %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%\n",
    "\n",
    "    import tkinter as tk\n",
    "    import csv\n",
    "    import tempfile\n",
    "    from pymongo import MongoClient\n",
    "    from tkinter import ttk\n",
    "    import subprocess\n",
    "\n",
    "    # Rest of the code remains the same...\n",
    "\n",
    "\n",
    "    def display_records(self):\n",
    "        # Connect to MongoDB\n",
    "        client = MongoClient('mongodb://localhost:27017')\n",
    "        db = client['breast']\n",
    "        collection2 = db['clinical_data']\n",
    "\n",
    "        # Perform aggregate pipeline to match females and sort by Age_InitDiag\n",
    "        pipeline = [\n",
    "            {\"$match\": {\"Gender\": \"FEMALE\"}},\n",
    "            {\"$sort\": {\"Age_InitDiag\": -1}}\n",
    "        ]\n",
    "        result = collection2.aggregate(pipeline)\n",
    "\n",
    "        # Create a temporary file to save the CSV data\n",
    "        temp_file_path = tempfile.NamedTemporaryFile(mode='w', delete=False, suffix='.csv').name\n",
    "\n",
    "        # Save the records as a CSV file\n",
    "        with open(temp_file_path, mode='w', newline='') as file:\n",
    "            writer = csv.writer(file)\n",
    "            headers = [\"RefSeq_accession_number\", \"ID\", \"Gender\", \"ER\", \"PR\", \"Node\", \"Node_Coded\",\n",
    "                       \"Metastasis\", \"Metastasis_Coded\", \"survive\", \"PAM50_mRNA\", \"methylation_clust\",\n",
    "                       \"gebe_exp_clust\", \"Age_InitDiag\", \"HER2_FinalStatus\", \"survive_code\", \"OS_Time\"]\n",
    "            writer.writerow(headers)\n",
    "            for record in result:\n",
    "                writer.writerow([record.get(key, \"\") for key in headers])\n",
    "\n",
    "        # Open the CSV file using the default system application\n",
    "        subprocess.run(['xdg-open', temp_file_path])\n",
    "        \n",
    "#------------\n",
    "#----cli----------######################### retrive one################################\n",
    "\n",
    "    def retrieve_document_cli(self):\n",
    "        \n",
    "            accession_number = self.label_entry_doc.get()\n",
    "\n",
    "            # Perform the query\n",
    "            document = self.collection2.find_one({'RefSeq_accession_number': accession_number})\n",
    "\n",
    "            if document:\n",
    "                # Display the document in a message box\n",
    "                messagebox.showinfo(\"Document\", str(document))\n",
    "            else:\n",
    "                messagebox.showinfo(\"Error\", \"Document not found.\")\n",
    "                \n",
    "\n",
    "\n",
    "#----cli-----------%%%%%%%%%%%%%%%%%%%%%%% not equal age=68 %%%%%%%%%%%%%%%%%5\n",
    "    def search_records(self):\n",
    "        # Connect to MongoDB\n",
    "        client = MongoClient('mongodb://localhost:27017/')\n",
    "        db = client['breast']\n",
    "        collection2 = db['clinical_data']\n",
    "\n",
    "        # Perform the MongoDB query\n",
    "        query = {\"Age_InitDiag\": {\"$ne\": 68}}\n",
    "        results = list(collection2.find(query))\n",
    "\n",
    "        # Create a new window\n",
    "        window = tk.Toplevel()\n",
    "        window.title(\"Search Results\")\n",
    "\n",
    "        # Get the screen width and height\n",
    "        screen_width = window.winfo_screenwidth()\n",
    "        screen_height = window.winfo_screenheight()\n",
    "\n",
    "        # Create a table frame to hold the table and scrollbar\n",
    "        table_frame = ttk.Frame(window)\n",
    "        table_frame.pack(fill=\"both\", expand=True)\n",
    "\n",
    "        # Create a table to display the results\n",
    "        table = ttk.Treeview(table_frame)\n",
    "        table[\"columns\"] = list(results[0].keys())\n",
    "\n",
    "        # Format the table columns\n",
    "        for column in table[\"columns\"]:\n",
    "            table.column(column, width=100)\n",
    "            table.heading(column, text=column)\n",
    "\n",
    "        # Insert the data into the table\n",
    "        for row in results:\n",
    "            table.insert(\"\", \"end\", values=list(row.values()))\n",
    "\n",
    "        # Add a vertical scrollbar to the table\n",
    "        scrollbar_y = ttk.Scrollbar(table_frame, orient=\"vertical\", command=table.yview)\n",
    "        table.configure(yscroll=scrollbar_y.set)\n",
    "        scrollbar_y.pack(side=\"right\", fill=\"y\")\n",
    "\n",
    "        # Add a horizontal scrollbar to the table\n",
    "        scrollbar_x = ttk.Scrollbar(window, orient=\"horizontal\", command=table.xview)\n",
    "        table.configure(xscroll=scrollbar_x.set)\n",
    "        scrollbar_x.pack(side=\"bottom\", fill=\"x\")\n",
    "\n",
    "        # Pack the table\n",
    "        table.pack(fill=\"both\", expand=True)\n",
    "\n",
    "        # Set the window size to occupy the full screen\n",
    "        window.geometry(f\"{screen_width}x{screen_height}\")\n",
    "#---cil function ################### sort <1> ascending ##########################################################3\n",
    "\n",
    "    def display_records_one(self):\n",
    "        # Connect to MongoDB\n",
    "        client = MongoClient('mongodb://localhost:27017')\n",
    "        db = client['breast']\n",
    "        collection2 = db['clinical_data']\n",
    "\n",
    "        # Perform aggregate pipeline to match females and sort by Age_InitDiag\n",
    "        pipeline = [\n",
    "            {\"$match\": {\"Gender\": \"FEMALE\"}},\n",
    "            {\"$sort\": {\"Age_InitDiag\": 1}}\n",
    "        ]\n",
    "        result = collection2.aggregate(pipeline)\n",
    "\n",
    "        # Create a temporary file to save the CSV data\n",
    "        temp_file_path = tempfile.NamedTemporaryFile(mode='w', delete=False, suffix='.csv').name\n",
    "\n",
    "        # Save the records as a CSV file\n",
    "        with open(temp_file_path, mode='w', newline='') as file:\n",
    "            writer = csv.writer(file)\n",
    "            headers = [\"RefSeq_accession_number\", \"ID\", \"Gender\", \"ER\", \"PR\", \"Node\", \"Node_Coded\",\n",
    "                       \"Metastasis\", \"Metastasis_Coded\", \"survive\", \"PAM50_mRNA\", \"methylation_clust\",\n",
    "                       \"gebe_exp_clust\", \"Age_InitDiag\", \"HER2_FinalStatus\", \"survive_code\", \"OS_Time\"]\n",
    "            writer.writerow(headers)\n",
    "            for record in result:\n",
    "                writer.writerow([record.get(key, \"\") for key in headers])\n",
    "\n",
    "        # Open the CSV file using the default system application\n",
    "        subprocess.run(['xdg-open', temp_file_path])\n",
    "\n",
    "#---cil function ################### OR ##########################################################3\n",
    "    def retrieve_or_documents(self):\n",
    "        field1 = self.field1_entry.get()\n",
    "        value1 = self.value1_entry.get()\n",
    "        field2 = self.field2_entry.get()\n",
    "        value2 = self.value2_entry.get()\n",
    "\n",
    "        # Connect to the MongoDB server\n",
    "        client = MongoClient('localhost', 27017)\n",
    "\n",
    "        # Access the database and collection\n",
    "        db = client['breast']\n",
    "        collection = db['clinical_data']\n",
    "\n",
    "        # Prepare the queries\n",
    "        query1 = { field1: value1 }\n",
    "        query2 = { field2: value2 }\n",
    "        query = { '$or': [query1, query2] }\n",
    "\n",
    "        # Retrieve the documents matching the query\n",
    "        documents = list(collection.find(query))\n",
    "\n",
    "        # Create the GUI window\n",
    "        window = tk.Tk()\n",
    "        window.title('Clinical Data Results')\n",
    "\n",
    "        # Create the treeview widget\n",
    "        tree = ttk.Treeview(window)\n",
    "\n",
    "        # Define the columns\n",
    "        tree['columns'] = tuple(documents[0].keys()) if documents else ()\n",
    "\n",
    "        # Format the columns\n",
    "        for column in tree['columns']:\n",
    "            tree.column(column, anchor='center')\n",
    "            tree.heading(column, text=column)\n",
    "\n",
    "        # Insert the document data into the treeview\n",
    "        for doc in documents:\n",
    "            tree.insert('', 'end', values=list(doc.values()))\n",
    "\n",
    "        # Add scrollbar\n",
    "        scrollbar = ttk.Scrollbar(window, orient=\"vertical\", command=tree.yview)\n",
    "        tree.configure(yscroll=scrollbar.set)\n",
    "        scrollbar.pack(side=\"right\", fill=\"y\")\n",
    "\n",
    "        # Pack the treeview\n",
    "        tree.pack(expand=True, fill='both')\n",
    "#--cli_ function ################### update_document_clinical#############################33\n",
    "    def update_document_cli(self):\n",
    "        client = MongoClient()\n",
    "\n",
    "    # Access the specified database and collection\n",
    "        db = client['breast']\n",
    "        collection2 = db['clinical_data']\n",
    "        \n",
    "            # Retrieve data from the collection\n",
    "        accession_number = self.ID_entry.get()\n",
    "        new_survive = self.survive_entry.get()\n",
    "        \n",
    "            # Update the document object based on the filter and update queries\n",
    "            \n",
    "        filter_query = {'RefSeq_accession_number': accession_number} #//// sure id_entry\n",
    "        update_query = {'$set': {'survive': new_survive}}\n",
    "        collection2.update_one(filter_query, update_query)\n",
    "        messagebox.showinfo('success', 'update success !')\n",
    "#---cli- FUNCTION  2 %%%%%%%%%%%%%%%%%%%%%%%%%%%%%% group --min %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%\n",
    "    def show_min(self):\n",
    "        client = MongoClient()\n",
    "\n",
    "    # Access the specified database and collection\n",
    "        db = client['breast']\n",
    "        collection2 = db['clinical_data']\n",
    "        result_window = tk.Tk()\n",
    "        result_window.title(\"Result\")\n",
    "\n",
    "\n",
    "\n",
    "        # Perform the aggregation query\n",
    "        result = collection2.aggregate([{\"$group\": {\"_id\": \"$Gender\", \"Min\": {\"$min\": \"$Age_InitDiag\"}}}])\n",
    "\n",
    "        # Create a new window to display the result\n",
    "\n",
    "        # Create a label to show the result\n",
    "        result_label = tk.Label(result_window, text=\"\", width=30)\n",
    "        result_label.pack(padx=10, pady=10)\n",
    "\n",
    "        # Format the result and display it in the label\n",
    "        text = \"\"\n",
    "        for doc in result:\n",
    "            text += f\"{doc['_id']}: {doc['Min']}\\n\"\n",
    "        result_label.config(text=text if text else \"No data available\")\n",
    "#---cli- FUNCTION   %%%%%%%%%%%%%%%%%%%%%%%%%%%%%% group --max %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%\n",
    "    def show_max(self):\n",
    "        client = MongoClient()\n",
    "\n",
    "    # Access the specified database and collection\n",
    "        db = client['breast']\n",
    "        collection2 = db['clinical_data']\n",
    "        result_window = tk.Tk()\n",
    "        result_window.title(\"Result\")\n",
    "\n",
    "\n",
    "\n",
    "        # Perform the aggregation query\n",
    "        result = collection2.aggregate([\n",
    "            {\"$group\": {\"_id\": \"$Gender\", \"Max\": {\"$max\": \"$Age_InitDiag\"}}}\n",
    "        ])\n",
    "\n",
    "        # Create a new window to display the result\n",
    "\n",
    "        # Create a label to show the result\n",
    "        result_label = tk.Label(result_window, text=\"\", width=30)\n",
    "        result_label.pack(padx=10, pady=10)\n",
    "\n",
    "        # Format the result and display it in the label\n",
    "        text = \"\"\n",
    "        for doc in result:\n",
    "            text += f\"{doc['_id']}: {doc['Max']}\\n\"\n",
    "        result_label.config(text=text if text else \"No data available\")\n",
    "\n",
    "#----cli function ############################refrence to clinical data ############################\n",
    "    def retrieve_clinical_data(self):\n",
    "            refseq_acc = self.refrence_entry.get()\n",
    "\n",
    "            # Connect to MongoDB\n",
    "            client = MongoClient('mongodb://localhost:27017/')\n",
    "            db = client['breast']  # Replace 'your_database_name' with your actual database name\n",
    "            cancer_collection = db['cancerf']\n",
    "            clinical_data_collection = db['clinical_data']\n",
    "\n",
    "            # Find the document in the 'cancer' collection with the specified RefSeq Accession Number\n",
    "            cancer_doc = cancer_collection.find_one({\"RefSeq_accession_number\": refseq_acc})\n",
    "\n",
    "            if cancer_doc:\n",
    "                # Get the RefSeq Accession Number from the 'cancer' collection document\n",
    "                refseq_acc_cancer = cancer_doc['RefSeq_accession_number']\n",
    "\n",
    "                # Find the document in the 'clinical_data' collection with the matching RefSeq Accession Number\n",
    "                clinical_data_doc = clinical_data_collection.find_one({\"RefSeq_accession_number\": refseq_acc_cancer})\n",
    "\n",
    "            if clinical_data_doc:\n",
    "                # Display the document in a message box\n",
    "                messagebox.showinfo(\"Document\", str(clinical_data_doc))\n",
    "            else:\n",
    "                messagebox.showinfo(\"Error\", \"Document not found.\")\n",
    "\n",
    "        #---show all databases %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%5\n",
    "    def show_databases(self):\n",
    "        # Connect to MongoDB\n",
    "        client = MongoClient('mongodb://localhost:27017/')\n",
    "\n",
    "        # Get the list of database names\n",
    "        database_names = client.list_database_names()\n",
    "\n",
    "        # Display the list of databases\n",
    "        messagebox.showinfo(\"MongoDB Databases\", \"\\n\".join(database_names))\n",
    "\n",
    "####################################################  $and###############\n",
    "    def retrieve_documents(self):\n",
    "        field1 = self.field1_entry.get()\n",
    "        value1 = self.value1_entry.get()\n",
    "        field2 = self.field2_entry.get()\n",
    "        value2 = self.value2_entry.get()\n",
    "\n",
    "        # Connect to the MongoDB server\n",
    "        client = MongoClient('localhost', 27017)\n",
    "\n",
    "        # Access the database and collection\n",
    "        db = client['breast']\n",
    "        collection = db['clinical_data']\n",
    "\n",
    "        # Prepare the $and query\n",
    "        query = {\n",
    "            '$and': [\n",
    "                {field1: value1},\n",
    "                {field2: value2}\n",
    "            ]\n",
    "        }\n",
    "\n",
    "        # Retrieve the documents matching the query\n",
    "        documents = list(collection.find(query))\n",
    "\n",
    "        # Create the GUI window\n",
    "        window = tk.Tk()\n",
    "        window.title('Clinical Data Results')\n",
    "\n",
    "        # Create the treeview widget\n",
    "        tree = ttk.Treeview(window)\n",
    "\n",
    "        # Define the columns\n",
    "        tree['columns'] = tuple(documents[0].keys()) if documents else ()\n",
    "\n",
    "        # Format the columns\n",
    "        for column in tree['columns']:\n",
    "            tree.column(column, anchor='center')\n",
    "            tree.heading(column, text=column)\n",
    "\n",
    "        # Insert the document data into the treeview\n",
    "        for doc in documents:\n",
    "            tree.insert('', 'end', values=list(doc.values()))\n",
    "\n",
    "        # Add scrollbar\n",
    "        scrollbar = ttk.Scrollbar(window, orient=\"vertical\", command=tree.yview)\n",
    "        tree.configure(yscroll=scrollbar.set)\n",
    "        scrollbar.pack(side=\"right\", fill=\"y\")\n",
    "\n",
    "        # Pack the treeview\n",
    "        tree.pack(expand=True, fill='both')\n",
    "    \n",
    "    #---cli_ function ################################## limit###########################################################\n",
    "\n",
    "\n",
    "    def limit_documents(self):\n",
    "            # Get the number of documents to retrieve from the user\n",
    "            limit = int(self.limit_entry.get())\n",
    "\n",
    "          # Connect to MongoDB\n",
    "            client = MongoClient('mongodb://localhost:27017/')\n",
    "            db = client['breast']  # Replace 'your_database_name' with your actual database name\n",
    "            collection = db['clinical_data']\n",
    "\n",
    "            # Retrieve the documents based on the specified limit\n",
    "            documents = collection.find().limit(limit)\n",
    "         # Create a new window to display the result\n",
    "            result_window = tk.Toplevel()\n",
    "            result_window.title(\"Documents\")\n",
    "            result_window.geometry(\"800x400\")\n",
    "\n",
    "            # Create a Treeview widget to display the result in a table-like structure\n",
    "            tree = ttk.Treeview(result_window)\n",
    "            tree[\"columns\"] = (\"RefSeq Accession Number\", \"ID\", \"Gender\", \"ER\", \"PR\", \"Node\", \"Node_Coded\", \"Metastasis\", \"Metastasis_Coded\", \"Survive\", \"PAM50 mRNA\", \"Methylation Clust\", \"GEBE Exp Clust\", \"Age InitDiag\", \"HER2 FinalStatus\", \"Survive Code\", \"OS Time\")\n",
    "            tree.heading(\"RefSeq Accession Number\", text=\"RefSeq Accession Number\")\n",
    "            tree.heading(\"ID\", text=\"ID\")\n",
    "            tree.heading(\"Gender\", text=\"Gender\")\n",
    "            tree.heading(\"ER\", text=\"ER\")\n",
    "            tree.heading(\"PR\", text=\"PR\")\n",
    "            tree.heading(\"Node\", text=\"Node\")\n",
    "            tree.heading(\"Node_Coded\", text=\"Node Coded\")\n",
    "            tree.heading(\"Metastasis\", text=\"Metastasis\")\n",
    "            tree.heading(\"Metastasis_Coded\", text=\"Metastasis Coded\")\n",
    "            tree.heading(\"Survive\", text=\"Survive\")\n",
    "            tree.heading(\"PAM50 mRNA\", text=\"PAM50 mRNA\")\n",
    "            tree.heading(\"Methylation Clust\", text=\"Methylation Clust\")\n",
    "            tree.heading(\"GEBE Exp Clust\", text=\"GEBE Exp Clust\")\n",
    "            tree.heading(\"Age InitDiag\", text=\"Age InitDiag\")\n",
    "            tree.heading(\"HER2 FinalStatus\", text=\"HER2 FinalStatus\")\n",
    "            tree.heading(\"Survive Code\", text=\"Survive Code\")\n",
    "            tree.heading(\"OS Time\", text=\"OS Time\")\n",
    "\n",
    "            # Create a scrollbar\n",
    "            scrollbar = ttk.Scrollbar(result_window, orient=\"vertical\", command=tree.yview)\n",
    "            scrollbar.pack(side=\"right\", fill=\"y\")\n",
    "\n",
    "            tree.configure(yscrollcommand=scrollbar.set)\n",
    "\n",
    "            # Insert the documents into the Treeview widget\n",
    "            for document in documents:\n",
    "                refseq = document['RefSeq_accession_number']\n",
    "                id_val = document['ID']\n",
    "                gender = document['Gender']\n",
    "                er = document['ER']\n",
    "                pr = document['PR']\n",
    "                node = document['Node']\n",
    "                node_coded = document['Node_Coded']\n",
    "                metastasis = document['Metastasis']\n",
    "                metastasis_coded = document['Metastasis_Coded']\n",
    "                survive = document['survive']\n",
    "                pam50_mrna = document['PAM50_mRNA']\n",
    "                methylation_clust = document['methylation_clust']\n",
    "                gebe_exp_clust = document['gebe_exp_clust']\n",
    "                age_initdiag = document['Age_InitDiag']\n",
    "                her2_finalstatus = document['HER2_FinalStatus']\n",
    "                survive_code = document['survive_code']\n",
    "                os_time = document['OS_Time']\n",
    "\n",
    "                tree.insert(\"\", \"end\", values=(refseq, id_val, gender, er, pr, node, node_coded, metastasis, metastasis_coded, survive, pam50_mrna, methylation_clust, gebe_exp_clust, age_initdiag, her2_finalstatus, survive_code, os_time))\n",
    "\n",
    "            # Pack the Treeview widget\n",
    "            tree.pack(fill=\"both\", expand=True)\n",
    "#---cli- FUNCTION  2 %%%%%%%%%%%%%%%%%%%%%%%%%%%%%% retrieve 2 col %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%\n",
    "\n",
    "    def retrieve_columns(self):\n",
    "            # Get the user input\n",
    "            col1 = self.col1_entry.get()\n",
    "            col2 = self.col2_entry.get()\n",
    "\n",
    "            # Connect to MongoDB\n",
    "            client = MongoClient('mongodb://localhost:27017/')\n",
    "            db = client['breast']\n",
    "            collection = db['clinical_data']\n",
    "\n",
    "            # Retrieve all documents from the collection\n",
    "            documents = collection.find()\n",
    "\n",
    "            # Create a new window to display the result\n",
    "            result_window = tk.Toplevel()\n",
    "            result_window.title(\"Column Data\")\n",
    "            result_window.geometry(\"600x400\")\n",
    "\n",
    "            # Create a Treeview widget to display the result in a table-like structure\n",
    "            tree = ttk.Treeview(result_window)\n",
    "            tree[\"columns\"] = (col1, col2)\n",
    "            tree.heading(col1, text=col1)\n",
    "            tree.heading(col2, text=col2)\n",
    "\n",
    "            # Create a scrollbar\n",
    "            scrollbar = ttk.Scrollbar(result_window, orient=\"vertical\", command=tree.yview)\n",
    "            scrollbar.pack(side=\"right\", fill=\"y\")\n",
    "\n",
    "            tree.configure(yscrollcommand=scrollbar.set)\n",
    "\n",
    "            # Insert the column values into the Treeview widget\n",
    "            for document in documents:\n",
    "                value1 = document.get(col1, \"\")\n",
    "                value2 = document.get(col2, \"\")\n",
    "                tree.insert(\"\", \"end\", values=(value1, value2))\n",
    "\n",
    "            # Pack the Treeview widget\n",
    "            tree.pack(fill=\"both\", expand=True)\n",
    "\n",
    "            \n",
    "#---cli- FUNCTION   %%%%%%%%%%%%%%%%%%%%%%%%%%%%%% limit3 min %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%\n",
    "            \n",
    "            \n",
    "    def limit_min(self):\n",
    "        # Connect to the MongoDB server\n",
    "        client = MongoClient('mongodb://localhost:27017')\n",
    "        # Access the desired database and collection\n",
    "        db = client['breast']\n",
    "        collection2 = db['clinical_data']\n",
    "\n",
    "        # Execute the aggregation pipeline\n",
    "        pipeline = [\n",
    "            {\"$group\": {\"_id\": \"$Gender\", \"Min\": {\"$min\": \"$Age_InitDiag\"}}},\n",
    "            {\"$limit\": 3}\n",
    "        ]\n",
    "        result = list(collection2.aggregate(pipeline))\n",
    "\n",
    "        # Create a new window for displaying the result\n",
    "        result_window = tk.Toplevel()\n",
    "\n",
    "        # Create a CSV file to store the result\n",
    "        with open('result.csv', 'w', newline='') as csvfile:\n",
    "            writer = csv.DictWriter(csvfile, fieldnames=['Gender', 'Min'])\n",
    "            writer.writeheader()\n",
    "\n",
    "            # Write the result to the CSV file and display it in the window\n",
    "            for record in result:\n",
    "                writer.writerow({'Gender': record['_id'], 'Min': record['Min']})\n",
    "                tk.Label(result_window, text=f\"Gender: {record['_id']}, Min: {record['Min']}\").pack()\n",
    "\n",
    "#---cli func %%%%%%%%%%%%%%%%%%%%%%%%%%%%% $ sum operator in survive_code to count dead $$$$$$$$$$$$$$$$$4\n",
    "\n",
    "    def calculate_sum(self):\n",
    "        # Connect to MongoDB\n",
    "        client = MongoClient('mongodb://localhost:27017/')\n",
    "        db = client['breast']\n",
    "        collection2 = db['clinical_data']\n",
    "\n",
    "        # Perform the MongoDB aggregation\n",
    "        pipeline = [\n",
    "            {\n",
    "                \"$match\": {\n",
    "                    \"survive\": \"DECEASED\"\n",
    "                }\n",
    "            },\n",
    "            {\n",
    "                \"$group\": {\n",
    "                    \"_id\": \"RefSeq_accession_number\",\n",
    "                    \"sum\": {\"$sum\": \"$survive_code\"}\n",
    "                }\n",
    "            }\n",
    "        ]\n",
    "        result = list(collection2.aggregate(pipeline))\n",
    "\n",
    "        # Extract the sum from the aggregation result\n",
    "        sum_value = result[0][\"sum\"]\n",
    "\n",
    "        # Display the message box\n",
    "        messagebox.showinfo(\"Sum of DECEASED Patients\", f\"The sum of all DECEASED patients is: {sum_value}\")\n",
    "\n",
    "#---cli Function %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%max limit 5 %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%\n",
    "\n",
    "    def limit_max(self):\n",
    "        # Connect to the MongoDB server\n",
    "        client = MongoClient('mongodb://localhost:27017')\n",
    "        # Access the desired database and collection\n",
    "        db = client['breast']\n",
    "        collection2 = db['clinical_data']\n",
    "\n",
    "        # Execute the aggregation pipeline\n",
    "        pipeline = [\n",
    "            {\"$group\": {\"_id\": \"$Gender\", \"Max\": {\"$max\": \"$Age_InitDiag\"}}},\n",
    "            {\"$limit\": 5}\n",
    "        ]\n",
    "        result = list(collection2.aggregate(pipeline))\n",
    "\n",
    "        # Create a new window for displaying the result\n",
    "        result_window = tk.Toplevel()\n",
    "\n",
    "        # Create a CSV file to store the result\n",
    "        with open('result.csv', 'w', newline='') as csvfile:\n",
    "            writer = csv.DictWriter(csvfile, fieldnames=['Gender', 'Min'])\n",
    "            writer.writeheader()\n",
    "\n",
    "            # Write the result to the CSV file and display it in the window\n",
    "            for record in result:\n",
    "                writer.writerow({'Gender': record['_id'], 'Min': record['Min']})\n",
    "                tk.Label(result_window, text=f\"Gender: {record['_id']}, Min: {record['Min']}\").pack()\n",
    "\n",
    "#---cli- FUNCTION  %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%  update_Many_documents ################################33\n",
    "    def update_Many_documents(self):\n",
    "          # Connect to your MongoDB database and collection\n",
    "            self.client = MongoClient('mongodb://localhost:27017/')\n",
    "            self.db = self.client['breast']\n",
    "            self.collection = self.db['clinical_data']\n",
    "\n",
    "            field = self.field_entry.get()\n",
    "            old_value = self.old_value_entry.get()\n",
    "            new_value = self.new_value_entry.get()\n",
    "\n",
    "            # Update all documents with the given field value\n",
    "            result = self.collection.update_many({field: old_value}, {\"$set\": {field: new_value}})\n",
    "\n",
    "            if result.modified_count > 0:\n",
    "                messagebox.showinfo('Success', 'Documents updated successfully.')\n",
    "            else:\n",
    "                messagebox.showinfo('Error', 'No documents found with the specified field value or no updates were made.')\n",
    "#######3\n",
    "\n",
    "    def get_indexes():\n",
    "        # Connect to the MongoDB database\n",
    "        client = MongoClient('mongodb://localhost:27017')\n",
    "        db = client['breast']\n",
    "        collection = db['clinical_data']\n",
    "\n",
    "        # Get the indexes of the collection\n",
    "        indexes = collection.index_information()\n",
    "\n",
    "        # Clear existing data in the labels\n",
    "        index_info_label.config(text=\"\")\n",
    "\n",
    "        # Insert index data into the label\n",
    "        for index_name, index_info in indexes.items():\n",
    "            index_info_label.config(text=index_info_label.cget(\"text\") + f\"{index_name}: {index_info}\\n\")\n",
    "\n",
    "\n",
    "#---cli- FUNCTION  2 %%%%%%%%%%%%%%%%%%%%%%%%%%%%%% retrieve %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%\n",
    "    def retrieve_data_cli(self):\n",
    "        # Retrieve data from the collection\n",
    "        data = self.collection2.find().sort('_id', DESCENDING)\n",
    "\n",
    "        # Create a new window for the results\n",
    "        results_window_cli = tk.Toplevel(self.hello)\n",
    "        results_window_cli.title(\"Results\")\n",
    "        results_window_cli.geometry(\"800x600\")\n",
    "\n",
    "        # Create a frame to hold the table and scrollbars\n",
    "        frame = tk.Frame(results_window_cli)\n",
    "        frame.pack(fill='both', expand=True)\n",
    "\n",
    "        # Create a vertical scrollbar\n",
    "        y_scrollbar = ttk.Scrollbar(frame, orient='vertical')\n",
    "        y_scrollbar.pack(side='right', fill='y')\n",
    "\n",
    "        # Create a horizontal scrollbar\n",
    "        x_scrollbar = ttk.Scrollbar(frame, orient='horizontal')\n",
    "        x_scrollbar.pack(side='bottom', fill='x')\n",
    "\n",
    "        # Create a treeview widget to display the data\n",
    "        treeview = ttk.Treeview(frame, yscrollcommand=y_scrollbar.set, xscrollcommand=x_scrollbar.set)\n",
    "        treeview.pack(fill='both', expand=True)\n",
    "\n",
    "        # Configure the scrollbars\n",
    "        y_scrollbar.configure(command=treeview.yview)\n",
    "        x_scrollbar.configure(command=treeview.xview)\n",
    "\n",
    "        # Define the columns\n",
    "        columns = tuple(data[0].keys())\n",
    "        treeview['columns'] = columns\n",
    "\n",
    "        # Configure the columns\n",
    "        for column in columns:\n",
    "            treeview.column(column, anchor='center', width=100)\n",
    "            treeview.heading(column, text=column)\n",
    "\n",
    "        # Add a border style to the treeview\n",
    "        style = ttk.Style()\n",
    "        style.configure(\"Treeview\", borderwidth=1, relief=\"solid\")\n",
    "\n",
    "        # Insert the data into the treeview\n",
    "        for document in data:\n",
    "            treeview.insert('', 'end', values=tuple(document.values()))\n",
    "\n",
    "        results_window_cli.mainloop()\n",
    "\n",
    "\n",
    "    \n",
    "if __name__ == \"__main__\":\n",
    "    app = App()\n",
    "    app.startpage()\n",
    "    #################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "17cc56fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "import tkinter as tk\n",
    "from tkinter import ttk\n",
    "from pymongo import MongoClient\n",
    "\n",
    "def get_indexes():\n",
    "    # Connect to the MongoDB database\n",
    "    client = MongoClient('mongodb://localhost:27017')\n",
    "    db = client['breast']\n",
    "    collection = db['clinical_data']\n",
    "\n",
    "    # Get the indexes of the collection\n",
    "    indexes = collection.index_information()\n",
    "\n",
    "    # Clear existing data in the labels\n",
    "    index_info_label.config(text=\"\")\n",
    "    \n",
    "    # Insert index data into the label\n",
    "    for index_name, index_info in indexes.items():\n",
    "        index_info_label.config(text=index_info_label.cget(\"text\") + f\"{index_name}: {index_info}\\n\")\n",
    "\n",
    "# Create a Tkinter window\n",
    "root = tk.Tk()\n",
    "root.title(\"Indexing App\")\n",
    "\n",
    "# Create a label to display index information\n",
    "index_info_label = tk.Label(root, text=\"\", justify=\"left\", font=(\"Arial\", 12))\n",
    "index_info_label.pack(pady=10)\n",
    "\n",
    "# Create a button to get the indexes\n",
    "indexes_button = tk.Button(root, text=\"Get Indexes\", command=get_indexes)\n",
    "indexes_button.pack()\n",
    "\n",
    "# Start the Tkinter event loop\n",
    "root.mainloop()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3148b584",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
